{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Bad key \"text.kerning_factor\" on line 4 in\n",
      "/Users/sudeshnaghosh/anaconda3/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test_patch.mplstyle.\n",
      "You probably need to get an updated matplotlibrc file from\n",
      "https://github.com/matplotlib/matplotlib/blob/v3.1.3/matplotlibrc.template\n",
      "or from the matplotlib source distribution\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from os.path import join\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "from scipy.stats import binom_test\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nImports and merges NanoARG Full Table and assembly information (length and read-count).\\nCalculates quantities and relative abundances of scaffolds with ARGs, MRGs, MGEs and \\npathogen scaffols. \\n\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Imports and merges NanoARG Full Table and assembly information (length and read-count).\n",
    "Calculates quantities and relative abundances of scaffolds with ARGs, MRGs, MGEs and \n",
    "pathogen scaffols. \n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_overlap_not_ARG_strand_selection(df, \n",
    "                                            max_overlap=250, \n",
    "                                            verbose=False, \n",
    "                                            strand_selection='no'):\n",
    "    \"\"\" \n",
    "    Rearrange open reading frames according to position.\n",
    "    Find open reading frames identified as ARGs that overlap with other \n",
    "    (MGEs, MRGs, Functional Genes) open reading frames and \n",
    "    remove them. \n",
    "    It is necessary to run this before many of the other functions.\n",
    "    For example, the same region identified as ARG and MGE can result in \n",
    "    misidentification of contigs with both ARG and MGE. Overlapping \n",
    "    open reading frame identifications also make it more difficult to analyze \n",
    "    contiguous regions.\n",
    "    \n",
    "    Example\n",
    "    -------\n",
    "          \n",
    "    s1 ----ARGs------- e1 \n",
    "         s2 --------- e2 removed if overlap > max_overlap\n",
    "                  s3-------------e3 keep if overlap < max_overlap\n",
    "                           s4 ------ e4 \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        data frame with NanoARG table\n",
    "    max_overlap : int\n",
    "        maximum allowed overlap between consequitive orfs\n",
    "    \n",
    "    strand_selection : string\n",
    "        choose 'no','yes'\n",
    "        \n",
    "        default run with 'no', preferred for subsequent data analysis\n",
    "        ignore +/- strand\n",
    "        (11) scaffold_10 4879-6615 ARGs\n",
    "           (12) 5947-6558  removing Functional Genes! overlap 611\n",
    "        (27) scaffold_10 44937-45410 ARGs\n",
    "        (30) scaffold_10 47395-49008 ARGs\n",
    "           (31) 47395-48969  removing MRGs! overlap 1574\n",
    "           (32) 47443-48996  removing Functional Genes! overlap 1553\n",
    "        (36) scaffold_10 52079-52978 ARGs\n",
    "        (37) scaffold_10 56260-57171 ARGs\n",
    "           (38) 56260-57168  removing Functional Genes! overlap 908\n",
    "        (42) scaffold_10002 34-753 ARGs\n",
    "           (43) 40-768  removing Functional Genes! overlap 713\n",
    "           \n",
    "        run with 'yes'\n",
    "        only remove if orfs have same 5'-> 3' direction\n",
    "        \n",
    "        (11) scaffold_10 4879-6615 ARGs\n",
    "           (12) 5947-6558 strand + + removing Functional Genes! overlap 611\n",
    "        (27) scaffold_10 44937-45410 ARGs\n",
    "        (30) scaffold_10 47395-49008 ARGs\n",
    "           (31) 47395-48969 strand + - overlap 1574\n",
    "           (32) 47443-48996 strand + - overlap 1553\n",
    "        (36) scaffold_10 52079-52978 ARGs\n",
    "        (37) scaffold_10 56260-57171 ARGs\n",
    "           (38) 56260-57168 strand + - overlap 908\n",
    "        (42) scaffold_10002 34-753 ARGs\n",
    "           (43) 40-768 strand + + removing Functional Genes! overlap 713\n",
    "        \n",
    "\n",
    "    verbose : bool\n",
    "        print debug info\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    Dataframe with entries removed\n",
    "    \n",
    "    \n",
    "   \n",
    "    \"\"\"\n",
    "    if verbose: \n",
    "        print(\"sorting...\")\n",
    "        \n",
    "    group_column = 2\n",
    "\n",
    "    D_sorted = df.groupby(df.index).apply(\n",
    "        lambda x: x.sort_values('start', ascending=True))\n",
    "    \n",
    "    Ddrop = D_sorted.droplevel(level=0).copy()\n",
    "\n",
    "    if verbose: \n",
    "        print(\"searching\")\n",
    "\n",
    "    for i in range(len(Ddrop)-1):\n",
    "        if Ddrop.group.iloc[i] == 'ARGs':\n",
    "            scaffold = Ddrop.index[i]\n",
    "\n",
    "            if verbose: \n",
    "                print(\"({}) {} {}-{} {}\".format(\n",
    "                    i, scaffold, Ddrop.start.iloc[i], Ddrop.end.iloc[i], Ddrop.group.iloc[i]))\n",
    "\n",
    "            j = i + 1\n",
    "            while (j<len(Ddrop) and (Ddrop.start.iloc[j] <= Ddrop.end.iloc[i])\n",
    "                   and (Ddrop.index[j] == scaffold)):\n",
    "\n",
    "                if verbose:\n",
    "                    print(\"   ({}) {}-{} \".format(\n",
    "                        j, Ddrop.start.iloc[j], Ddrop.end.iloc[j]), end='')\n",
    "\n",
    "                overlap = np.min([Ddrop.end.iloc[j], Ddrop.end.iloc[i]]) - Ddrop.start.iloc[j]\n",
    "                if overlap >= max_overlap:\n",
    "                    if Ddrop.group.iloc[j] != 'ARGs':\n",
    "                        if strand_selection == 'yes':\n",
    "                            if verbose:\n",
    "                                print(\"strand {} {}\".format(\n",
    "                                    Ddrop.strand.iloc[i], Ddrop.strand.iloc[j]), end = '')\n",
    "                               \n",
    "                            if Ddrop.strand.iloc[j] == Ddrop.strand.iloc[i]:\n",
    "                                if verbose:\n",
    "                                    print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                                Ddrop.iat[j, group_column] = np.nan\n",
    "                        elif strand_selection == 'no':\n",
    "                            if verbose:\n",
    "                                print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                            Ddrop.iat[j, group_column] = np.nan\n",
    "                else:\n",
    "\n",
    "                    if verbose:\n",
    "                        print(\"keeping {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "\n",
    "                if verbose:\n",
    "                    print(\" overlap {}\".format(overlap))\n",
    "\n",
    "                j = j + 1      \n",
    "                \n",
    "    for i in range(1, len(Ddrop)):\n",
    "        if Ddrop.group.iloc[i] == 'ARGs':\n",
    "            scaffold = Ddrop.index[i]\n",
    "\n",
    "            if verbose: \n",
    "                print(\"({}) {} {}-{} {}\".format(\n",
    "                    i, scaffold, Ddrop.start.iloc[i], Ddrop.end.iloc[i], Ddrop.group.iloc[i]))\n",
    "\n",
    "     \n",
    "            j = i - 1\n",
    "            while (j>=0 and (Ddrop.start.iloc[i] <= Ddrop.end.iloc[j])\n",
    "                   and (Ddrop.index[j] == scaffold)):\n",
    "\n",
    "                if verbose:\n",
    "                    print(\"   ({}) {}-{} \".format(\n",
    "                        j, Ddrop.start.iloc[j], Ddrop.end.iloc[j]), end='')\n",
    "\n",
    "                overlap = np.min([Ddrop.end.iloc[j], Ddrop.end.iloc[i]]) - Ddrop.start.iloc[i]\n",
    "                if overlap >= max_overlap:\n",
    "                    if Ddrop.group.iloc[j] != 'ARGs':\n",
    "                        if strand_selection == 'yes':\n",
    "                            if verbose:\n",
    "                                print(\"strand {} {}\".format(\n",
    "                                    Ddrop.strand.iloc[i], Ddrop.strand.iloc[j]), end='')\n",
    "                            if Ddrop.strand.iloc[j] == Ddrop.strand.iloc[i]:\n",
    "                                if verbose:\n",
    "                                    print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                                Ddrop.iat[j, group_column] = np.nan\n",
    "                        elif strand_selection == 'no':\n",
    "                            if verbose:\n",
    "                                print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                            Ddrop.iat[j, group_column] = np.nan\n",
    "                    \n",
    "                else:\n",
    "\n",
    "                    if verbose:\n",
    "                        print(\"keeping {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "\n",
    "                if verbose:\n",
    "                    print(\" overlap {}\".format(overlap))\n",
    "\n",
    "                j = j - 1\n",
    "                \n",
    "    df_remove_overlap = Ddrop.dropna(subset=['group']) \n",
    "    \n",
    "    return df_remove_overlap       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Necessary to run 'remove_overlap_not_ARG_strand_selection' before. \n",
    "Calculates quantities and relative abundances of scaffolds with ARGs,\n",
    "MGEs and scaffolds identified as pathogens.\n",
    "\"\"\"\n",
    "\n",
    "def count_scaffolds_with_MGEs_and_ARGs_normalized(df, min_score=40000, \n",
    "                                                  min_length = 500):\n",
    "    \"\"\"\n",
    "    Count\n",
    "        ARG_counts\n",
    "        scaffold_ARGs\n",
    "        scaffold_ARGs_normalized\n",
    "        scaffold_MGEs\n",
    "        scaffold_MGEs_with_ARGs\n",
    "        scaffold_MGEs_with_ARGs_normalized\n",
    "        %_scaffold_MGEs_with_ARGs/scaffold_ARGs\n",
    "        scaffold_pathogens\n",
    "        scaffold_pathogens_with_args\n",
    "        scaffold_pathogens_with_mges\n",
    "        scaffold_pathogens_with_args_mges\n",
    "        scaffold_has_pathogens_ARGs_MGEs_normalized\n",
    "    \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold, ARGs and MGEs data\n",
    "    min_score : int\n",
    "        taxa_centrifuge_score cutoff for taxonomy classification\n",
    "    min_length : int\n",
    "        minimum length of scaffold to include in analysis\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dict:\n",
    "        dictionary with ratio and counts\n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df_length_filtered = df[df['length'] > min_length]\n",
    "    \n",
    "    scaffold_has_args_normalized = 0\n",
    "    scaffold_has_args=0\n",
    "    old_index_arg = 0\n",
    "    for i in range(len(df_length_filtered)):\n",
    "        if df_length_filtered.group.iloc[i] == 'ARGs':\n",
    "            if df_length_filtered.index[i] != old_index_arg:\n",
    "                scaffold_has_args += 1\n",
    "                scaffold_has_args_normalized  += (\n",
    "                    df_length_filtered.normalized_coverage.iloc[i])\n",
    "                old_index_arg = df_length_filtered.index[i]\n",
    "    \n",
    "    \n",
    "    scaffold_has_mges = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: (x['group']=='MGEs').any()\n",
    "    ).sum()\n",
    "    \n",
    "    scaffold_has_mges_with_args=0\n",
    "    scaffold_has_mges_with_args_norm=0\n",
    "    same_index=0\n",
    "    for scaffold, data in df_length_filtered.groupby(level=0):\n",
    "        groups = np.array(data.group)\n",
    "        normalized_coverage = np.array(data.normalized_coverage)\n",
    "        indices = np.array(data.index)\n",
    "        for i in range(len(groups)):\n",
    "            if (groups[i] == 'ARGs'):\n",
    "                for j in range(len(groups)):\n",
    "                    if (groups[j] == 'MGEs'):\n",
    "                        if indices[j] != same_index:\n",
    "                            scaffold_has_mges_with_args += 1\n",
    "                            scaffold_has_mges_with_args_norm += normalized_coverage[j]\n",
    "                            same_index = indices[j]\n",
    "        \n",
    "    \n",
    "    scaffold_has_pathogens = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: \n",
    "            (x['is_pathogen']==1).any() \n",
    "            and (x['taxa_centrifuge_score']>min_score).any()\n",
    "    ).sum()\n",
    "    \n",
    "    scaffold_has_pathogens_args = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: \n",
    "            (x['is_pathogen']==1).any() \n",
    "            and (x['taxa_centrifuge_score']>min_score).any() \n",
    "            and (x['group']=='ARGs').any()\n",
    "    ).sum()\n",
    "    \n",
    "    scaffold_has_pathogens_mges = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: \n",
    "            (x['is_pathogen']==1).any() \n",
    "            and (x['taxa_centrifuge_score']>min_score).any()\n",
    "            and (x['group']=='MGEs').any() \n",
    "    ).sum()\n",
    "    \n",
    "    df_length_taxa_filtered = df_length_filtered[df_length_filtered['taxa_centrifuge_score'] > min_score]\n",
    "    scaffold_has_pathogens_args_mges=0\n",
    "    scaffold_has_pathogens_args_mges_norm=0\n",
    "    same_index_pathogen=0\n",
    "    for scaffold, data in df_length_taxa_filtered.groupby(level=0):\n",
    "        groups = np.array(data.group)\n",
    "        normalized_coverage = np.array(data.normalized_coverage)\n",
    "        indices = np.array(data.index)\n",
    "        pathogens = np.array(data.is_pathogen)\n",
    "        for i in range(len(groups)):\n",
    "            if (pathogens[i] == 1):\n",
    "                for j in range(len(groups)):\n",
    "                    if (groups[j] == 'ARGs'):\n",
    "                        for k in range(len(groups)):\n",
    "                            if (groups[k] == 'MGEs'):\n",
    "                                if indices[k] != same_index_pathogen:\n",
    "                                    scaffold_has_pathogens_args_mges += 1\n",
    "                                    scaffold_has_pathogens_args_mges_norm += normalized_coverage[k]\n",
    "                                    same_index_pathogen = indices[k]\n",
    "    \n",
    "    ARG_counts = len(df_length_filtered[df_length_filtered['group'] == 'ARGs'])\n",
    "    \n",
    "    \n",
    "    return {\n",
    "        'ARG_counts' : ARG_counts,\n",
    "        'scaffold_ARGs': scaffold_has_args,\n",
    "        'scaffold_ARGs_normalized' : scaffold_has_args_normalized,\n",
    "        'scaffold_MGEs': scaffold_has_mges,\n",
    "        'scaffold_MGEs_with_ARGs': scaffold_has_mges_with_args,\n",
    "        'scaffold_MGEs_with_ARGs_normalized' :scaffold_has_mges_with_args_norm,\n",
    "        '%_scaffold_MGEs_with_ARGs/scaffold_ARGs': (\n",
    "            scaffold_has_mges_with_args / scaffold_has_args * 100\n",
    "        ),\n",
    "        'scaffold_pathogens' : scaffold_has_pathogens,\n",
    "        'scaffold_pathogens_with_ARGs' : scaffold_has_pathogens_args,\n",
    "        'scaffold_pathogens_with_MGEs' : scaffold_has_pathogens_mges,\n",
    "        'scaffold_pathogens_with_ARGs_MGEs' : scaffold_has_pathogens_args_mges,\n",
    "        'scaffold_has_pathogens_ARGs_MGEs_normalized' : scaffold_has_pathogens_args_mges_norm\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Returns quantity and relative abundance of desired ARG subtype scaffold. \n",
    "\"\"\"\n",
    "\n",
    "def ARG_subtype_count(df, min_bitscore=50, \n",
    "                      min_id_arg=25, min_coverage_arg=0.4, arg_name = 'sul1'):\n",
    "    \n",
    "    \"\"\"\n",
    "    Count\n",
    "        arg_subtype_count = count # of ARG subtype occurance\n",
    "        scaffold_arg_subtype = count # of scaffolds with ARG_subtype\n",
    "        scaffold_has_arg_subtype_normalized = scaffolds with ARG_subtype in RPKM\n",
    "    \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold, ARGs and MGEs data\n",
    "    min_bitscore : int\n",
    "        minimum bitscore, default is 50\n",
    "    min_id_arg : int\n",
    "        minimum identity of ARG, default is 25    \n",
    "    min_coverage_arg : int\n",
    "        minimum coverage of ARG, default is 0.4 in NanoARG\n",
    "    arg_name : name of ARG subtype to be counted\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dict:\n",
    "        dictionary with counts   \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    \n",
    "    df_bitscore_adjusted = df[df['bitscore'] > min_bitscore]\n",
    "    \n",
    "    df_bit_arg_adjusted = df_bitscore_adjusted.drop(df_bitscore_adjusted[ \n",
    "        (df_bitscore_adjusted['group'] == 'ARGs') \n",
    "        & ((df_bitscore_adjusted['identity'] < min_id_arg)\n",
    "            | (df_bitscore_adjusted['coverage'] < min_coverage_arg))].index)\n",
    "    \n",
    "    \n",
    "    arg_subtype_count = len(df_bit_arg_adjusted[df_bit_arg_adjusted['gene_name'] == arg_name])\n",
    "    \n",
    "    scaffold_has_arg_subtype = df_bit_arg_adjusted.groupby(level=0).apply(\n",
    "        lambda x: (x['gene_name']==arg_name).any()\n",
    "    ).sum()\n",
    "    \n",
    "    \n",
    "    scaffold_has_arg_subtype_normalized = 0\n",
    "    old_index = 0\n",
    "    for i in range(len(df_bit_arg_adjusted)):\n",
    "        if df_bit_arg_adjusted.gene_name.iloc[i] == arg_name:\n",
    "            if df_bit_arg_adjusted.index[i] != old_index:              \n",
    "                scaffold_has_arg_subtype_normalized  += (\n",
    "                    df_bit_arg_adjusted.normalized_coverage.iloc[i])\n",
    "                old_index = df_bit_arg_adjusted.index[i]\n",
    "                \n",
    "    return {\n",
    "        'arg_subtype_count' : arg_subtype_count,\n",
    "        'scaffold_arg_subtype' : scaffold_has_arg_subtype,\n",
    "        'scaffold_has_arg_subtype_normalized' : scaffold_has_arg_subtype_normalized\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_scaffolds(df, min_length=500):\n",
    "    \"\"\"    \n",
    "    Count number of scaffolds longer than min_length \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold length and read-count\n",
    "    min_length : int\n",
    "        minimum length of scaffold to include in analysis\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dict:\n",
    "        dictionary with counts\n",
    "           \n",
    "    \"\"\"\n",
    "    \n",
    "    df_length_filtered = df[df['length'] > min_length]\n",
    "    scaffold_counts = len(df_length_filtered.groupby(level=0))\n",
    "    \n",
    "    return {\n",
    "        '# scaffolds' : scaffold_counts,\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "It is necessary to run 'remove_overlap_not_ARG_strand_selection' before.\n",
    "Finds likely association between ARG subtype and MGE.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def ARG_MGE_association(df, min_bitscore=50, min_id_arg=25, min_id_mge=25,\n",
    "                         f=0.01, max_fdr=0.05, min_coverage_arg=0.4,\n",
    "                       min_coverage_mge=0.4, min_length=500):\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold, ARGs and MGEs data\n",
    "    min_bitscore : int\n",
    "        minimum bitscore, default is 50\n",
    "    min_id_arg : int\n",
    "        minimum identity of ARG, default is 25    \n",
    "    min_id_mge : int\n",
    "        minimum identity of MGE, default is 25\n",
    "    min_coverage_arg : int\n",
    "        minimum coverage of ARG, default is 0.4 in NanoARG\n",
    "    min_coverage_mge : int\n",
    "        minimum coverage of MGE, default is 0.4 in NanoARG\n",
    "    f : float\n",
    "        ratio of scaffolds with MGEs and total number of scaffolds\n",
    "        set f = 0 to bypass binom test with FDR (Benjamini/Hochberg) correction.\n",
    "        Binom test checks if fraction of ARG subtype scaffold with MGEs \n",
    "        exceeds fraction of scaffolds with MGEs in the sample, in other words,\n",
    "        if ARG subtype is associated with an MGE more often than expected \n",
    "    max_fdr : float\n",
    "        maximum FDR value to include in output \n",
    "    min_length : int\n",
    "        minimum length of scaffold to include in analysis\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    fraction_genes_with_mge_test_count: pd.DataFrame()\n",
    "        Dataframe with ratio and counts\n",
    "        fdr : corrected p-value\n",
    "        fraction_genes_with_mge : ARG subtype scaffold with MGE / ARG subtype scaffold\n",
    "        count_genes : count of ARG subtype\n",
    "        \n",
    "    \n",
    "    \"\"\"\n",
    "    df_length_filtered = df[df['length'] > min_length]\n",
    "    \n",
    "    df_bitscore_adjusted = df_length_filtered[df_length_filtered['bitscore'] > min_bitscore]\n",
    "\n",
    "    df_bit_arg_adjusted = df_bitscore_adjusted.drop(df_bitscore_adjusted[ \n",
    "        (df_bitscore_adjusted['group'] == 'ARGs') \n",
    "        & ((df_bitscore_adjusted['identity'] < min_id_arg)\n",
    "            | (df_bitscore_adjusted['coverage'] < min_coverage_arg))].index)\n",
    "\n",
    "    df_adjusted = df_bit_arg_adjusted.drop(df_bit_arg_adjusted[ \n",
    "        (df_bit_arg_adjusted['group'] == 'MGEs') \n",
    "        & ((df_bit_arg_adjusted['identity'] < min_id_mge)\n",
    "            | (df_bit_arg_adjusted['coverage'] < min_coverage_mge))].index)\n",
    "    \n",
    "    D = df_adjusted[df_adjusted.group.isin(['ARGs','MGEs'])]\n",
    "    D_args = df_adjusted[df_adjusted.group.isin(['ARGs'])]\n",
    "    count_genes = Counter(D_args['gene_name'])\n",
    "    scaffolds = D_args.index.unique()\n",
    "\n",
    "    count_genes_with_mge = Counter()\n",
    "    for i, scaffold in enumerate(scaffolds):\n",
    "        print(\"{:.2f}%\".format((i+1)/len(scaffolds)*100.), end='\\r')\n",
    "\n",
    "        in_scaffold = D[D.index==scaffold]\n",
    "        if 'MGEs' in D[D.index==scaffold]['group'].values:\n",
    "\n",
    "            for gene in count_genes:\n",
    "                if gene in in_scaffold['gene_name'].values:\n",
    "                    count_genes_with_mge[gene]+=1\n",
    "    print(\"           \", end=\"\\r\")\n",
    "\n",
    "    count_genes_once_per_scaffold = Counter()\n",
    "    for i, scaffold in enumerate(scaffolds):\n",
    "        print(\"{:.2f}%\".format((i+1)/len(scaffolds)*100.), end='\\r')\n",
    "        in_scaffold = D[D.index==scaffold]\n",
    "        for gene in count_genes:\n",
    "            if gene in in_scaffold['gene_name'].values:\n",
    "                count_genes_once_per_scaffold[gene]+=1\n",
    "    \n",
    "    \n",
    "\n",
    "    fraction_genes_with_mge = {}\n",
    "    for gene in count_genes_with_mge:\n",
    "        fraction_genes_with_mge[gene]=count_genes_with_mge[gene]/count_genes_once_per_scaffold[gene]\n",
    "    \n",
    "    fraction_genes_with_mge_df = pd.DataFrame.from_dict(fraction_genes_with_mge, orient='index')\n",
    "    \n",
    "    fraction_genes_with_mge_df = fraction_genes_with_mge_df.rename(columns={0: \"fraction_genes_with_mge\"})\n",
    "    \n",
    "    count_genes_df = pd.DataFrame.from_dict(count_genes, orient='index')\n",
    "    count_genes_df = count_genes_df.rename(columns={0: \"count_genes\"})\n",
    "\n",
    "    test_enriched = {}\n",
    "    for gene, count in count_genes_with_mge.items(): \n",
    "        test_enriched[gene] = binom_test(count, n=count_genes_once_per_scaffold[gene], p=f, alternative='greater')\n",
    "    \n",
    "    test_enriched = pd.Series(test_enriched)\n",
    "    \n",
    "    rejected, pvalue_corrected = (\n",
    "        fdrcorrection(test_enriched, alpha=max_fdr, method='indep', is_sorted=False))\n",
    "\n",
    "    test_enriched_fdr = {\n",
    "        'rejected': rejected.tolist(), \n",
    "        'fdr': pvalue_corrected.tolist(), \n",
    "        'gene_name': test_enriched.index.tolist()\n",
    "    }\n",
    "    test_enriched_fdr_df = pd.DataFrame(test_enriched_fdr)\n",
    "    test_enriched_fdr_df.index = test_enriched_fdr_df['gene_name']\n",
    "\n",
    "    fraction_genes_with_mge_test = (\n",
    "        pd.concat([test_enriched_fdr_df, fraction_genes_with_mge_df], axis=1, join='inner'))\n",
    "    \n",
    "    fraction_genes_with_mge_test = (\n",
    "        fraction_genes_with_mge_test.sort_values(by=['fraction_genes_with_mge'], ascending=False))\n",
    "    \n",
    "    fraction_genes_with_mge_test_count = (\n",
    "        pd.concat([fraction_genes_with_mge_test, count_genes_df], axis=1, join='inner'))\n",
    "    \n",
    "    fraction_genes_with_mge_test_count = (\n",
    "        fraction_genes_with_mge_test_count[fraction_genes_with_mge_test_count['rejected']])\n",
    "    \n",
    "    fraction_genes_with_mge_test_count.drop(['rejected','gene_name'], axis=1, inplace=True)\n",
    "    return fraction_genes_with_mge_test_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "It is necessary to run 'remove_overlap_not_ARG_strand_selection' before.\n",
    "Returns stretches of scaffolds with ARGs and MGEs.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def neighbor_stretch(df, min_ARG_count=1, max_distance=5000, \n",
    "                     number_neighboring_orf=4, \n",
    "                     sliding_window_step_increase=4, min_bitscore=50, \n",
    "                     min_id_arg=25, min_id_mge=25, min_coverage_arg=0.4, \n",
    "                     min_coverage_mge=0.4, verbose=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    Search for MGE and specified number of ARGs in a desired\n",
    "    stretch of scaffold window. This window slides forward at a \n",
    "    specified step.\n",
    "    Reports a dataframe of neighboring orfs with MGEs and ARGs.\n",
    "    \n",
    "    Exp. number_neighboring_orf=4, sliding_window_step_increase=1\n",
    "    max_distance = maximum distance between s and e\n",
    "    min_ARG_count=1\n",
    "    \n",
    "    s1-ARG--->  <-----   ------> --MGE-->e1    report\n",
    "              s2<-----   ------> --MGE--> <------e2    no ARG, do not report  \n",
    "              \n",
    "    \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold, ARGs, MGEs and start end \n",
    "        gene data\n",
    "    \n",
    "    max_distance : int\n",
    "        maximum length of scaffold segment with \n",
    "        neighboring MGEs and ARGs to explore\n",
    "        \n",
    "    min_ARG_count : int\n",
    "        minimum number of ARGs in the scaffold segment with \n",
    "        neighboring MGEs and ARGs\n",
    "        \n",
    "    number_neighboring_orf : int\n",
    "        maximum number of neighboring orfs to consider\n",
    "        \n",
    "    sliding_window_step_increase : int\n",
    "        int between 1 and number_neighboring_orf\n",
    "    \n",
    "    min_bitscore : int\n",
    "        minimum bitscore, default is 50\n",
    "        \n",
    "    min_id_arg : int\n",
    "        minimum aa identity of ARG, default is 25\n",
    "        \n",
    "    min_id_mge : int\n",
    "        minimum aa identity of MGE, default is 25\n",
    "        \n",
    "    min_coverage_arg : int\n",
    "        minimum coverage of ARG, default is 0.4 in NanoARG\n",
    "        \n",
    "    min_coverage_mge : int\n",
    "        minimum coverage of MGE, default is 0.4 in NanoARG\n",
    "        \n",
    "    verbose : bool\n",
    "        print progress and debug information\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    neighbor_genes : dataframe\n",
    "        dataframe with neighboring ARGs and MGEs\n",
    "    \n",
    "    \"\"\"\n",
    "    neighbor_genes = {}\n",
    "    \n",
    "    output_columns = ['gene_id', 'gene_name', 'group', 'category', 'start', 'end', 'strand',\n",
    "       'identity', 'bitscore', 'evalue', 'NCBI_taxa_id',\n",
    "       'taxa_centrifuge_score', 'species', 'coverage', 'is_pathogen']\n",
    "    \n",
    "    for column in output_columns:\n",
    "        neighbor_genes[column] = []\n",
    "        \n",
    "    \n",
    "    neighbor_genes['read'] = []\n",
    "\n",
    "    df_bitscore_adjusted = df[df['bitscore'] > min_bitscore]\n",
    "    \n",
    "    df_bit_arg_adjusted = df_bitscore_adjusted.drop(df_bitscore_adjusted[ \n",
    "        (df_bitscore_adjusted['group'] == 'ARGs') \n",
    "        & ((df_bitscore_adjusted['identity'] < min_id_arg)\n",
    "            | (df_bitscore_adjusted['coverage'] < min_coverage_arg))].index)\n",
    "\n",
    "    df_adjusted = df_bit_arg_adjusted.drop(df_bit_arg_adjusted[ \n",
    "        (df_bit_arg_adjusted['group'] == 'MGEs') \n",
    "        & ((df_bit_arg_adjusted['identity'] < min_id_mge)\n",
    "            | (df_bit_arg_adjusted['coverage'] < min_coverage_mge))].index)\n",
    "    \n",
    "    \n",
    "    if verbose:\n",
    "        print(\"\\ndf1:\", df_bitscore_adjusted)\n",
    "        print(\"\\ndf2:\", df_bit_arg_adjusted)\n",
    "        print(\"\\ndf3:\", df_adjusted)\n",
    "    \n",
    "    \n",
    "    for scaffold, data in df_adjusted.groupby(level=0):\n",
    "        starts = np.array(data.start)\n",
    "        ends = np.array(data.end)\n",
    "        groups = np.array(data.group)\n",
    "        \n",
    "        if verbose:\n",
    "            print(\"-----------------\")\n",
    "            print(\"\\nscaffold:\", scaffold)\n",
    "            print(\"\\nstarts:\", starts)\n",
    "            print(\"\\nends:\", ends)\n",
    "            print(\"\\ngroups:\", groups)\n",
    "            print(\"-----------------\")\n",
    "            print(len(groups))\n",
    "        \n",
    "        for i in range(0, len(groups)-number_neighboring_orf+1, sliding_window_step_increase):\n",
    "            max_neighbor_orfs = min(number_neighboring_orf, len(groups) - i)\n",
    "            count = Counter(groups[i:i+max_neighbor_orfs] == 'ARGs')\n",
    "            \n",
    "            \n",
    "            if count[True] >= min_ARG_count and (groups[i:i+max_neighbor_orfs] == 'MGEs').any():\n",
    "            \n",
    "                if (ends[i+max_neighbor_orfs-1] - starts[i]) < max_distance:\n",
    "                                \n",
    "                        neighbor_genes['read'] += [\n",
    "                            data.index.to_list()[i + d] for d in range(max_neighbor_orfs)\n",
    "                        ]\n",
    "                        \n",
    "                        for column in output_columns:\n",
    "                            neighbor_genes[column] += [\n",
    "                                data[column].to_list()[i+d] for d in range(max_neighbor_orfs)\n",
    "                            ]\n",
    "    neighbor_genes = pd.DataFrame(neighbor_genes)\n",
    "    neighbor_genes.index = neighbor_genes['read']\n",
    "    neighbor_genes = neighbor_genes.drop(['read'], axis =1)\n",
    "    \n",
    "    return neighbor_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCL0_W2\n",
      "HCL0_W3\n",
      "Before_GAC_1\n",
      "After_GAC_1\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "INPUT REQUIRED. IMPORTING NANOARG DATA.\n",
    "------------------------------------------------------\n",
    "\n",
    "Import multiple NanoARG Full Tables (.tsv) and save in a dictionary.\n",
    "\n",
    "Input required:\n",
    "1. Path to the directory with NanoARG full tables\n",
    "\n",
    "Output:\n",
    "1. Disctionary called all_full_tables, each sample is saved \n",
    "    in a different dataframe with sample name.\n",
    " \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "***provide PATH to the directory with files*********\n",
    "\n",
    "\"\"\"\n",
    "path = '/Users/sudeshnaghosh/Dropbox/VT/test2/'\n",
    "\n",
    "\n",
    "#------------------------------------------------\n",
    "\n",
    "\n",
    "DATA_DIR = path\n",
    "\n",
    "all_full_tables = {}\n",
    "\n",
    "for file in os.listdir(DATA_DIR):\n",
    "    df_name = file.split('.')[0]\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    all_full_tables[df_name] = pd.read_csv(DATA_DIR + file, sep=',', header=0, index_col=0)\n",
    "    \n",
    "#del all_full_tables[\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/After_GAC_1.fa\n",
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/Before_GAC_1.fa\n",
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/HCL0_W2.fa\n",
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/HCL0_W3.fa\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "INPUT REQUIRED. IMPORTING ASSEMBLY INFORMATION.\n",
    "------------------------------------------------------\n",
    "\n",
    "Extract length and read_count information about each scaffold \n",
    "in a sample and store as dataframe. Store dataframes in a dictionary.\n",
    "\n",
    "\n",
    "Input:\n",
    "\n",
    "Input file format, from IDBA-UD\n",
    "\n",
    ">scaffold_0 length_285165 read_count_39166\n",
    "CGGGCCCTACCGTAGCAGCCGCTACGGTAGGGCCCGTGTCCAGT......\n",
    ">scaffold_1 length_297291 read_count_43258\n",
    "GAGAGCGTCATCGATCATCGCGACGCAGGTTAACAGCGATTGCCGATGTTTACAACC........\n",
    "\n",
    "Output:\n",
    "    \n",
    "Dictionary called assembly\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "***provide PATH to the directory with IDBA-UD files**********\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "path = '/Users/sudeshnaghosh/Dropbox/VT/assembly_test'\n",
    "#---------------------------------------------------\n",
    "\n",
    "\n",
    "DATA_DIR = path\n",
    "\n",
    "assembly = {}\n",
    "\n",
    "\n",
    "for file in os.listdir(DATA_DIR):\n",
    "    if file.endswith(\".fa\"):\n",
    "        file_path = join(DATA_DIR, file)\n",
    "        print(file_path)\n",
    "        df_name = file.split('.')[0]\n",
    "        scaffold = []\n",
    "        length = []\n",
    "        read_count = []\n",
    "        \n",
    "        for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "            scaffold.append(record.description.split()[0])\n",
    "            length.append(record.description.split()[1].split('_')[1])\n",
    "            read_count.append(record.description.rsplit()[2].split('_')[2])\n",
    "            \n",
    "            \n",
    "        assembly[df_name] = pd.DataFrame({\n",
    "            'read':scaffold,\n",
    "            'length':length,\n",
    "            'read_count':read_count\n",
    "        })\n",
    "        assembly[df_name].set_index('read', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before_GAC_1\n",
      "After_GAC_1\n",
      "HCL0_W3\n",
      "HCL0_W2\n"
     ]
    }
   ],
   "source": [
    "\"\"\"IMPORTING COVERAGE FILES\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "data format (scaffold,coverage)\n",
    "\n",
    "\n",
    "\n",
    "genome,12.541209956374683\n",
    "scaffold_0,23.00672351625\n",
    "scaffold_1,94.66073208748007\n",
    "scaffold_10,22.6907473287\n",
    "scaffold_100,17.560885991499994\n",
    "scaffold_1000,23.2361095814\n",
    "scaffold_10000,7.281462299999999\n",
    "scaffold_100000,5.385235400000001\n",
    "scaffold_100001,2.77849415\n",
    "scaffold_100002,3.1011257399999996\n",
    "scaffold_100003,4.524878\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "path = '/Users/sudeshnaghosh/Dropbox/coverage_test/'\n",
    "\n",
    "\n",
    "#------------------------------------------------\n",
    "\n",
    "\n",
    "DATA_DIR = path\n",
    "\n",
    "coverage = {}\n",
    "\n",
    "for file in os.listdir(DATA_DIR):\n",
    "    df_name = file.split('.')[0].rsplit('_',1)[0]\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    coverage[df_name] = pd.read_csv(DATA_DIR + file, sep=',', header=None, index_col=0)\n",
    "    \n",
    "#del all_full_tables[\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_0</th>\n",
       "      <td>491867</td>\n",
       "      <td>60073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_1</th>\n",
       "      <td>412542</td>\n",
       "      <td>44554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_2</th>\n",
       "      <td>224903</td>\n",
       "      <td>21783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_3</th>\n",
       "      <td>195468</td>\n",
       "      <td>291957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_4</th>\n",
       "      <td>319504</td>\n",
       "      <td>57709</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            length read_count\n",
       "read                         \n",
       "scaffold_0  491867      60073\n",
       "scaffold_1  412542      44554\n",
       "scaffold_2  224903      21783\n",
       "scaffold_3  195468     291957\n",
       "scaffold_4  319504      57709"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly['HCL0_W3'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coverage['LCM5_W3'].iloc[0,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before_GAC_1\n",
      "After_GAC_1\n",
      "HCL0_W3\n",
      "HCL0_W2\n"
     ]
    }
   ],
   "source": [
    "for name, keys in coverage.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    coverage[df_name] = coverage[df_name].astype(float)\n",
    "    coverage[df_name]['all_scaffold'] = coverage[df_name].iloc[0,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"\"\"\n",
    "# For each scaffold, calculate rpm and rpkm \n",
    "# -----------------------------------------\n",
    "\n",
    "\n",
    "# For each scaffold, calculate rpm and rpkm and add to the disctionary with \n",
    "# length and read_count.\n",
    "\n",
    "# normalized_read_count (rpm) = \n",
    "# read_count of scaffold / (total # of read_counts in a sample / 1,000,000)\n",
    "# normalized_read_count_per_kb (rpkm) = \n",
    "# normalized_read_count (rpm) / length of scaffold in kb\n",
    "# normalized_length (bases per million) =\n",
    "# (length of scaffold / (length summation of scaffolds / 1,000,000))\n",
    "\n",
    "# read_count_per_length (or 'read_count_per_base') = \n",
    "# (read_count of scaffold / (total # of read_counts in a sample / 1,000,000))\n",
    "# /\n",
    "# (length of scaffold / (length summation of scaffolds / 1,000,000))\n",
    "\n",
    "# The updated dataframes are stored in the dictionary called assembly.\n",
    "\n",
    "\n",
    "# \"\"\"\n",
    "\n",
    "\n",
    "# for name, keys in assembly.items():    \n",
    "#     df_name = name\n",
    "#     # printing sample name indicates progress\n",
    "#     print(df_name)\n",
    "#     SUM = ((assembly[df_name]['read_count']).astype(float)).sum()\n",
    "#     assembly[df_name] = assembly[df_name].astype(float)\n",
    "#     assembly[df_name]['normalized_read_count'] = (\n",
    "#         (assembly[df_name]['read_count'] * 1000000) / SUM)\n",
    "#     assembly[df_name]['normalized_read_count_per_kb'] = (\n",
    "#         (assembly[df_name]['normalized_read_count'] * 1000)/\n",
    "#         assembly[df_name]['length'])\n",
    "#     assembly[df_name]['length_sum'] = assembly[df_name]['length'].sum()\n",
    "#     assembly[df_name]['normalized_length'] = (\n",
    "#         (assembly[df_name]['length'] * 1000000) / assembly[df_name]['length_sum'])\n",
    "#     assembly[df_name]['read_count_per_length'] = (\n",
    "#         assembly[df_name]['normalized_read_count']/\n",
    "#         assembly[df_name]['normalized_length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_0</th>\n",
       "      <td>709020</td>\n",
       "      <td>97243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_1</th>\n",
       "      <td>551495</td>\n",
       "      <td>312878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_2</th>\n",
       "      <td>954535</td>\n",
       "      <td>669486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_3</th>\n",
       "      <td>303743</td>\n",
       "      <td>175890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_4</th>\n",
       "      <td>302455</td>\n",
       "      <td>32823</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            length read_count\n",
       "read                         \n",
       "scaffold_0  709020      97243\n",
       "scaffold_1  551495     312878\n",
       "scaffold_2  954535     669486\n",
       "scaffold_3  303743     175890\n",
       "scaffold_4  302455      32823"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly['After_GAC_1'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "COMBINING ASSEMBLY INFORMATION WITH COVERAGE\n",
    "--------------------------------------------\n",
    "\n",
    "Add scaffold length, read_count and coverage.\n",
    "Store dataframes in a dictionary called assembly_coverage.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "assembly_coverage = {}\n",
    "\n",
    "for name, keys in assembly.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    assembly_coverage[df_name] = (\n",
    "        pd.concat([assembly[df_name], coverage[df_name]], \n",
    "                  axis=1, sort=False, join='inner'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "      <th>1</th>\n",
       "      <th>all_scaffold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_0</th>\n",
       "      <td>491867</td>\n",
       "      <td>60073</td>\n",
       "      <td>20.876421</td>\n",
       "      <td>22.253067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_1</th>\n",
       "      <td>412542</td>\n",
       "      <td>44554</td>\n",
       "      <td>18.579756</td>\n",
       "      <td>22.253067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_2</th>\n",
       "      <td>224903</td>\n",
       "      <td>21783</td>\n",
       "      <td>16.761661</td>\n",
       "      <td>22.253067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_3</th>\n",
       "      <td>195468</td>\n",
       "      <td>291957</td>\n",
       "      <td>250.051349</td>\n",
       "      <td>22.253067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_4</th>\n",
       "      <td>319504</td>\n",
       "      <td>57709</td>\n",
       "      <td>32.088281</td>\n",
       "      <td>22.253067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            length read_count           1  all_scaffold\n",
       "scaffold_0  491867      60073   20.876421     22.253067\n",
       "scaffold_1  412542      44554   18.579756     22.253067\n",
       "scaffold_2  224903      21783   16.761661     22.253067\n",
       "scaffold_3  195468     291957  250.051349     22.253067\n",
       "scaffold_4  319504      57709   32.088281     22.253067"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_coverage['HCL0_W3'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "for name, keys in assembly_coverage.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    assembly_coverage[df_name] = assembly_coverage[df_name].astype(float)\n",
    "    assembly_coverage[df_name]['sum_assembled_read'] = assembly_coverage[df_name]['read_count'].sum()\n",
    "    assembly_coverage[df_name]['length_sum'] = assembly_coverage[df_name]['length'].sum()\n",
    "    assembly_coverage[df_name]['coverage_sum'] = assembly_coverage[df_name][1].sum()\n",
    "    assembly_coverage[df_name]['normalized_coverage'] = (\n",
    "        (assembly_coverage[df_name][1]/\n",
    "        assembly_coverage[df_name]['coverage_sum']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "      <th>1</th>\n",
       "      <th>all_scaffold</th>\n",
       "      <th>sum_assembled_read</th>\n",
       "      <th>length_sum</th>\n",
       "      <th>coverage_sum</th>\n",
       "      <th>normalized_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_0</th>\n",
       "      <td>491867.0</td>\n",
       "      <td>60073.0</td>\n",
       "      <td>20.876421</td>\n",
       "      <td>22.253067</td>\n",
       "      <td>25521993.0</td>\n",
       "      <td>235795745.0</td>\n",
       "      <td>2.716443e+06</td>\n",
       "      <td>0.000008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_1</th>\n",
       "      <td>412542.0</td>\n",
       "      <td>44554.0</td>\n",
       "      <td>18.579756</td>\n",
       "      <td>22.253067</td>\n",
       "      <td>25521993.0</td>\n",
       "      <td>235795745.0</td>\n",
       "      <td>2.716443e+06</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_2</th>\n",
       "      <td>224903.0</td>\n",
       "      <td>21783.0</td>\n",
       "      <td>16.761661</td>\n",
       "      <td>22.253067</td>\n",
       "      <td>25521993.0</td>\n",
       "      <td>235795745.0</td>\n",
       "      <td>2.716443e+06</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_3</th>\n",
       "      <td>195468.0</td>\n",
       "      <td>291957.0</td>\n",
       "      <td>250.051349</td>\n",
       "      <td>22.253067</td>\n",
       "      <td>25521993.0</td>\n",
       "      <td>235795745.0</td>\n",
       "      <td>2.716443e+06</td>\n",
       "      <td>0.000092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_4</th>\n",
       "      <td>319504.0</td>\n",
       "      <td>57709.0</td>\n",
       "      <td>32.088281</td>\n",
       "      <td>22.253067</td>\n",
       "      <td>25521993.0</td>\n",
       "      <td>235795745.0</td>\n",
       "      <td>2.716443e+06</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              length  read_count           1  all_scaffold  \\\n",
       "scaffold_0  491867.0     60073.0   20.876421     22.253067   \n",
       "scaffold_1  412542.0     44554.0   18.579756     22.253067   \n",
       "scaffold_2  224903.0     21783.0   16.761661     22.253067   \n",
       "scaffold_3  195468.0    291957.0  250.051349     22.253067   \n",
       "scaffold_4  319504.0     57709.0   32.088281     22.253067   \n",
       "\n",
       "            sum_assembled_read   length_sum  coverage_sum  normalized_coverage  \n",
       "scaffold_0          25521993.0  235795745.0  2.716443e+06             0.000008  \n",
       "scaffold_1          25521993.0  235795745.0  2.716443e+06             0.000007  \n",
       "scaffold_2          25521993.0  235795745.0  2.716443e+06             0.000006  \n",
       "scaffold_3          25521993.0  235795745.0  2.716443e+06             0.000092  \n",
       "scaffold_4          25521993.0  235795745.0  2.716443e+06             0.000012  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_coverage['HCL0_W3'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCL0_W2\n",
      "HCL0_W3\n",
      "Before_GAC_1\n",
      "After_GAC_1\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "COMBINING NANOARG FULL TABLES AND ASSEMBLY INFORMATION\n",
    "------------------------------------------------------\n",
    "\n",
    "Add scaffold length, read_count and normalized scaffold read_count \n",
    "(rpm and rpkm) to NanoARG table.\n",
    "Store dataframes in a dictionary called all_full_tables_assembly_data.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "all_full_tables_assembly_data = {}\n",
    "\n",
    "for name, keys in all_full_tables.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    all_full_tables_assembly_data[df_name] = (\n",
    "        pd.concat([all_full_tables[df_name], assembly_coverage[df_name]], \n",
    "                  axis=1, sort=False, join='inner'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>group</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>identity</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>evalue</th>\n",
       "      <th>...</th>\n",
       "      <th>coverage</th>\n",
       "      <th>is_pathogen</th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "      <th>1</th>\n",
       "      <th>all_scaffold</th>\n",
       "      <th>sum_assembled_read</th>\n",
       "      <th>length_sum</th>\n",
       "      <th>coverage_sum</th>\n",
       "      <th>normalized_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_16508</th>\n",
       "      <td>WP_007306505.1</td>\n",
       "      <td>IS1634 family transposase [Crocosphaera watsonii]</td>\n",
       "      <td>MGEs</td>\n",
       "      <td>transposase</td>\n",
       "      <td>1058</td>\n",
       "      <td>1459</td>\n",
       "      <td>+</td>\n",
       "      <td>35.1</td>\n",
       "      <td>102.4</td>\n",
       "      <td>6.500000e-21</td>\n",
       "      <td>...</td>\n",
       "      <td>0.724324</td>\n",
       "      <td>0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>6.709122</td>\n",
       "      <td>12.54121</td>\n",
       "      <td>14992547.0</td>\n",
       "      <td>259033383.0</td>\n",
       "      <td>2.552888e+06</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_9236</th>\n",
       "      <td>YP_002467958</td>\n",
       "      <td>PBP-1B</td>\n",
       "      <td>ARGs</td>\n",
       "      <td>beta-lactam</td>\n",
       "      <td>326</td>\n",
       "      <td>1816</td>\n",
       "      <td>+</td>\n",
       "      <td>26.6</td>\n",
       "      <td>137.9</td>\n",
       "      <td>1.300000e-32</td>\n",
       "      <td>...</td>\n",
       "      <td>0.718421</td>\n",
       "      <td>0</td>\n",
       "      <td>2392.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>6.479508</td>\n",
       "      <td>12.54121</td>\n",
       "      <td>14992547.0</td>\n",
       "      <td>259033383.0</td>\n",
       "      <td>2.552888e+06</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_9236</th>\n",
       "      <td>UniRef90_Q07259</td>\n",
       "      <td>Cluster: Putative transglycosylase H16 A0665</td>\n",
       "      <td>Functional Genes</td>\n",
       "      <td>Cluster: Putative transglycosylase H16 A0665</td>\n",
       "      <td>395</td>\n",
       "      <td>862</td>\n",
       "      <td>+</td>\n",
       "      <td>37.2</td>\n",
       "      <td>107.8</td>\n",
       "      <td>1.500000e-22</td>\n",
       "      <td>...</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0</td>\n",
       "      <td>2392.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>6.479508</td>\n",
       "      <td>12.54121</td>\n",
       "      <td>14992547.0</td>\n",
       "      <td>259033383.0</td>\n",
       "      <td>2.552888e+06</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_16501</th>\n",
       "      <td>BAC0162</td>\n",
       "      <td>Gallium</td>\n",
       "      <td>MRGs</td>\n",
       "      <td>Gallium</td>\n",
       "      <td>15</td>\n",
       "      <td>659</td>\n",
       "      <td>+</td>\n",
       "      <td>31.0</td>\n",
       "      <td>102.8</td>\n",
       "      <td>9.400000e-24</td>\n",
       "      <td>...</td>\n",
       "      <td>0.606742</td>\n",
       "      <td>0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>4.414802</td>\n",
       "      <td>12.54121</td>\n",
       "      <td>14992547.0</td>\n",
       "      <td>259033383.0</td>\n",
       "      <td>2.552888e+06</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_16501</th>\n",
       "      <td>BAC0604</td>\n",
       "      <td>Molybdenum</td>\n",
       "      <td>MRGs</td>\n",
       "      <td>Molybdenum</td>\n",
       "      <td>778</td>\n",
       "      <td>1440</td>\n",
       "      <td>+</td>\n",
       "      <td>30.0</td>\n",
       "      <td>102.8</td>\n",
       "      <td>9.500000e-24</td>\n",
       "      <td>...</td>\n",
       "      <td>0.648256</td>\n",
       "      <td>0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>4.414802</td>\n",
       "      <td>12.54121</td>\n",
       "      <td>14992547.0</td>\n",
       "      <td>259033383.0</td>\n",
       "      <td>2.552888e+06</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id  \\\n",
       "scaffold_16508   WP_007306505.1   \n",
       "scaffold_9236      YP_002467958   \n",
       "scaffold_9236   UniRef90_Q07259   \n",
       "scaffold_16501          BAC0162   \n",
       "scaffold_16501          BAC0604   \n",
       "\n",
       "                                                        gene_name  \\\n",
       "scaffold_16508  IS1634 family transposase [Crocosphaera watsonii]   \n",
       "scaffold_9236                                              PBP-1B   \n",
       "scaffold_9236        Cluster: Putative transglycosylase H16 A0665   \n",
       "scaffold_16501                                            Gallium   \n",
       "scaffold_16501                                         Molybdenum   \n",
       "\n",
       "                           group  \\\n",
       "scaffold_16508              MGEs   \n",
       "scaffold_9236               ARGs   \n",
       "scaffold_9236   Functional Genes   \n",
       "scaffold_16501              MRGs   \n",
       "scaffold_16501              MRGs   \n",
       "\n",
       "                                                    category  start   end  \\\n",
       "scaffold_16508                                   transposase   1058  1459   \n",
       "scaffold_9236                                    beta-lactam    326  1816   \n",
       "scaffold_9236   Cluster: Putative transglycosylase H16 A0665    395   862   \n",
       "scaffold_16501                                       Gallium     15   659   \n",
       "scaffold_16501                                    Molybdenum    778  1440   \n",
       "\n",
       "               strand  identity  bitscore        evalue  ...  coverage  \\\n",
       "scaffold_16508      +      35.1     102.4  6.500000e-21  ...  0.724324   \n",
       "scaffold_9236       +      26.6     137.9  1.300000e-32  ...  0.718421   \n",
       "scaffold_9236       +      37.2     107.8  1.500000e-22  ...  0.675325   \n",
       "scaffold_16501      +      31.0     102.8  9.400000e-24  ...  0.606742   \n",
       "scaffold_16501      +      30.0     102.8  9.500000e-24  ...  0.648256   \n",
       "\n",
       "                is_pathogen  length  read_count         1  all_scaffold  \\\n",
       "scaffold_16508            0  1743.0        74.0  6.709122      12.54121   \n",
       "scaffold_9236             0  2392.0        81.0  6.479508      12.54121   \n",
       "scaffold_9236             0  2392.0        81.0  6.479508      12.54121   \n",
       "scaffold_16501            0  1743.0        41.0  4.414802      12.54121   \n",
       "scaffold_16501            0  1743.0        41.0  4.414802      12.54121   \n",
       "\n",
       "                sum_assembled_read   length_sum  coverage_sum  \\\n",
       "scaffold_16508          14992547.0  259033383.0  2.552888e+06   \n",
       "scaffold_9236           14992547.0  259033383.0  2.552888e+06   \n",
       "scaffold_9236           14992547.0  259033383.0  2.552888e+06   \n",
       "scaffold_16501          14992547.0  259033383.0  2.552888e+06   \n",
       "scaffold_16501          14992547.0  259033383.0  2.552888e+06   \n",
       "\n",
       "                normalized_coverage  \n",
       "scaffold_16508             0.000003  \n",
       "scaffold_9236              0.000003  \n",
       "scaffold_9236              0.000003  \n",
       "scaffold_16501             0.000002  \n",
       "scaffold_16501             0.000002  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_full_tables_assembly_data['After_GAC_1'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nSAVE DICTIONARY\\n---------------\\nsave dictionary in all_full_tables_assembly_data_test.p\\n\\ndictionary can be loaded in new notebook for analysis\\n\\nwith open('all_full_tables_remove_overlap_test.p', 'rb') as fp:\\n    all_full_tables_remove_overlap = pickle.load(fp)\\n\""
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "SAVE DICTIONARY\n",
    "---------------\n",
    "save dictionary in all_full_tables_assembly_data_test.p\n",
    "\n",
    "dictionary can be loaded in new notebook for analysis\n",
    "\n",
    "with open('all_full_tables_remove_overlap_test.p', 'rb') as fp:\n",
    "    all_full_tables_remove_overlap = pickle.load(fp)\n",
    "\"\"\"\n",
    "\n",
    "# with open('all_full_tables_assembly_data_average_coverage.p', 'wb') as fp:\n",
    "#     pickle.dump(all_full_tables_assembly_data, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCL0_W2\n",
      "HCL0_W3\n",
      "Before_GAC_1\n",
      "After_GAC_1\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Running 'remove_overlap_not_ARG_strand_selection'\n",
    "------------------------------------------------\n",
    "rearranging with ascending start\n",
    "and storing updated dataframes in \n",
    "a dictionary called\n",
    "all_full_tables_remove_overlap.\n",
    "\n",
    "Set the max_overlap to desired value.\n",
    "\n",
    "IMPORTANT: this step is necessary for running\n",
    "many of the other functions.\n",
    "\n",
    "\"\"\"\n",
    "all_full_tables_remove_overlap = {}\n",
    "for name, keys in all_full_tables_assembly_data.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    all_full_tables_remove_overlap[df_name] = (\n",
    "        remove_overlap_not_ARG_strand_selection(all_full_tables_assembly_data[df_name], \n",
    "                                            max_overlap=250, \n",
    "                                            strand_selection='no',\n",
    "                                            verbose = False))\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "SAVE DICTIONARY\n",
    "---------------\n",
    "\n",
    "Helpful if you have a lot of samples\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "# with open('all_full_tables_remove_overlap_average_coverage.p', 'wb') as fp:\n",
    "#     pickle.dump(all_full_tables_remove_overlap, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_0</th>\n",
       "      <td>709020</td>\n",
       "      <td>97243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_1</th>\n",
       "      <td>551495</td>\n",
       "      <td>312878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_2</th>\n",
       "      <td>954535</td>\n",
       "      <td>669486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_3</th>\n",
       "      <td>303743</td>\n",
       "      <td>175890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_4</th>\n",
       "      <td>302455</td>\n",
       "      <td>32823</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            length read_count\n",
       "read                         \n",
       "scaffold_0  709020      97243\n",
       "scaffold_1  551495     312878\n",
       "scaffold_2  954535     669486\n",
       "scaffold_3  303743     175890\n",
       "scaffold_4  302455      32823"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly['After_GAC_1'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "runs 'count_scaffolds'\n",
    "---------------------\n",
    "\n",
    "Returns dataframe with number of scaffolds in a sample.\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "scaffold_count = pd.DataFrame(dtype=float)\n",
    "for name, df in assembly_coverage.items(): \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    count = count_scaffolds(df, min_length = 1000)\n",
    "    count_series = pd.Series(count)\n",
    "    count_rename = count_series.rename(name)\n",
    "    \n",
    "    scaffold_count = scaffold_count.append(count_rename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># scaffolds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>After_GAC_1</th>\n",
       "      <td>46811.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Before_GAC_1</th>\n",
       "      <td>35063.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2</th>\n",
       "      <td>40216.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W3</th>\n",
       "      <td>45471.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              # scaffolds\n",
       "After_GAC_1       46811.0\n",
       "Before_GAC_1      35063.0\n",
       "HCL0_W2           40216.0\n",
       "HCL0_W3           45471.0"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scaffold_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCL0_W2\n",
      "HCL0_W3\n",
      "Before_GAC_1\n",
      "After_GAC_1\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "runs 'count_scaffolds_with_MGEs_and_ARGs_normalized'\n",
    "----------------------------------------------------\n",
    "\n",
    "Necessary to run 'remove_overlap_not_ARG_strand_selection' and 'count_scaffolds' before\n",
    "\n",
    "Estimate: 5 samples run in ~ 2 mins to run. More samples will take longer. \n",
    "\n",
    "\n",
    "Output\n",
    "------\n",
    "df : pd.dataframe\n",
    "    dataframe with counts, ratios and percent values\n",
    "    \n",
    "    ARG_counts : Number of ARGs in a sample \n",
    "    scaffold_ARGs : Number of scaffolds with at least one ARG \n",
    "    scaffold_ARGs_normalized : Scaffold with at least one ARG \n",
    "                               rel. abundance \n",
    "    scaffold_MGEs : Number of scaffolds with at least one MGE\n",
    "    scaffold_MGEs_with_ARGs : Number of scaffolds with at least one AGR \n",
    "                              and at least one MGE \n",
    "    scaffold_MGEs_with_ARGs_normalized : Scaffold with at least one each \n",
    "                                         of ARG and MGE rel.abundance\n",
    "    %_scaffold_MGEs_with_ARGs/scaffold_ARGs : (scaffold_MGEs_with_ARGs x 100)/scaffold_ARGs\n",
    "    scaffold_pathogens : Number of scaffolds identified as pathogen\n",
    "    scaffold_pathogens_with_ARGs : Number of scaffolds identified as pathogen with \n",
    "                                at least one ARG\n",
    "    scaffold_pathogens_with_MGEs : Number of scaffolds identified as pathogen with \n",
    "                                at least one MGE\n",
    "    scaffold_pathogens_with_ARGs_MGEs : Number of scaffolds identified \n",
    "                                        as pathogen with at least one AGR \n",
    "                                        and at least one MGE\n",
    "    scaffold_has_pathogens_ARGs_MGEs_normalized: Rel abundance of scaffolds identified \n",
    "                                        as pathogen with at least one AGR \n",
    "                                        and at least one MGE (rpkm)\n",
    "    # scaffolds :  Number of scaffolds in a sample\n",
    "    scaffold_MGEs/#scaffolds : scaffold_MGEs/# scaffolds\n",
    "    %scaffold_ARGs : (scaffold_ARGs  x 100) / # scaffolds\n",
    "    %scaffold_MGEs_with_ARGs : (scaffold_MGEs_with_ARGs  x 100) / # scaffolds \n",
    "    %pathogens : (scaffold_pathogens x 100) / # scaffolds \n",
    "    %pathogens_with_ARGs : (scaffold_pathogens_with_ARGs x 100) / # scaffolds\n",
    "    %pathogens_with_MGEs : (scaffold_pathogens_with_MGEs x 100) / # scaffolds\n",
    "    %pathogens_with_ARGs_MGEs : (scaffold_pathogens_with_ARGs_MGEs x 100) / # scaffolds \n",
    "    %scaffold_MGEs : (scaffold_MGEs x 100) / # scaffolds\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "columns = [\n",
    "           'ARG_counts', 'scaffold_ARGs',\n",
    "           'scaffold_ARGs_normalized',\n",
    "           'scaffold_MGEs',\n",
    "           'scaffold_MGEs_with_ARGs',\n",
    "           'scaffold_MGEs_with_ARGs_normalized',\n",
    "           '%_scaffold_MGEs_with_ARGs/scaffold_ARGs', 'scaffold_pathogens',\n",
    "           'scaffold_pathogens_with_ARGs', 'scaffold_pathogens_with_MGEs',\n",
    "           'scaffold_pathogens_with_ARGs_MGEs',\n",
    "           'scaffold_has_pathogens_ARGs_MGEs_normalized']\n",
    "scaffolds_with_MGEs_and_ARGs = pd.DataFrame(columns=columns, dtype=float)\n",
    "for name, df in all_full_tables_remove_overlap.items(): \n",
    "    print(name)\n",
    "    count = count_scaffolds_with_MGEs_and_ARGs_normalized(df, min_length = 0)\n",
    "    count_series = pd.Series(count)\n",
    "    count_rename = count_series.rename(name)\n",
    "    \n",
    "    scaffolds_with_MGEs_and_ARGs = scaffolds_with_MGEs_and_ARGs.append(count_rename)\n",
    "        \n",
    "\n",
    "scaffolds_with_MGEs_and_ARGs_store = scaffolds_with_MGEs_and_ARGs\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized = (\n",
    "    pd.concat([scaffolds_with_MGEs_and_ARGs_store, scaffold_count], \n",
    "              axis=1, join='inner'))\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs/#scaffolds']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs']/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%scaffold_ARGs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_ARGs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%scaffold_MGEs_with_ARGs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs_with_ARGs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens_with_ARGs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens_with_ARGs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens_with_MGEs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens_with_MGEs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens_with_ARGs_MGEs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens_with_ARGs_MGEs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%scaffold_MGEs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ARG_counts', 'scaffold_ARGs', 'scaffold_ARGs_normalized',\n",
       "       'scaffold_MGEs', 'scaffold_MGEs_with_ARGs',\n",
       "       'scaffold_MGEs_with_ARGs_normalized',\n",
       "       '%_scaffold_MGEs_with_ARGs/scaffold_ARGs', 'scaffold_pathogens',\n",
       "       'scaffold_pathogens_with_ARGs', 'scaffold_pathogens_with_MGEs',\n",
       "       'scaffold_pathogens_with_ARGs_MGEs',\n",
       "       'scaffold_has_pathogens_ARGs_MGEs_normalized', '# scaffolds',\n",
       "       'scaffold_MGEs/#scaffolds', '%scaffold_ARGs',\n",
       "       '%scaffold_MGEs_with_ARGs', '%pathogens', '%pathogens_with_ARGs',\n",
       "       '%pathogens_with_MGEs', '%pathogens_with_ARGs_MGEs', '%scaffold_MGEs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Prints result.\n",
    "Save this file and plot using desired software\n",
    "\"\"\"\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "MGEs_and_ARGs_pathogens_normalized['mean_cov_MGEs_with_ARGs/mean_cov'] = (\n",
    "(MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs_with_ARGs_normalized']*\n",
    "    MGEs_and_ARGs_pathogens_normalized['# scaffolds'])/\n",
    "    MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs_with_ARGs']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARG_counts</th>\n",
       "      <th>scaffold_ARGs</th>\n",
       "      <th>scaffold_ARGs_normalized</th>\n",
       "      <th>scaffold_MGEs</th>\n",
       "      <th>scaffold_MGEs_with_ARGs</th>\n",
       "      <th>scaffold_MGEs_with_ARGs_normalized</th>\n",
       "      <th>%_scaffold_MGEs_with_ARGs/scaffold_ARGs</th>\n",
       "      <th>scaffold_pathogens</th>\n",
       "      <th>scaffold_pathogens_with_ARGs</th>\n",
       "      <th>scaffold_pathogens_with_MGEs</th>\n",
       "      <th>...</th>\n",
       "      <th># scaffolds</th>\n",
       "      <th>scaffold_MGEs/#scaffolds</th>\n",
       "      <th>%scaffold_ARGs</th>\n",
       "      <th>%scaffold_MGEs_with_ARGs</th>\n",
       "      <th>%pathogens</th>\n",
       "      <th>%pathogens_with_ARGs</th>\n",
       "      <th>%pathogens_with_MGEs</th>\n",
       "      <th>%pathogens_with_ARGs_MGEs</th>\n",
       "      <th>%scaffold_MGEs</th>\n",
       "      <th>mean_cov_MGEs_with_ARGs/mean_cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCL0_W2</th>\n",
       "      <td>6679.0</td>\n",
       "      <td>5248.0</td>\n",
       "      <td>0.037242</td>\n",
       "      <td>3831.0</td>\n",
       "      <td>465.0</td>\n",
       "      <td>0.008308</td>\n",
       "      <td>8.860518</td>\n",
       "      <td>31.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>...</td>\n",
       "      <td>40216.0</td>\n",
       "      <td>0.095261</td>\n",
       "      <td>13.049533</td>\n",
       "      <td>1.156256</td>\n",
       "      <td>0.077084</td>\n",
       "      <td>0.017406</td>\n",
       "      <td>0.054705</td>\n",
       "      <td>0.004973</td>\n",
       "      <td>9.526059</td>\n",
       "      <td>0.718541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W3</th>\n",
       "      <td>7706.0</td>\n",
       "      <td>6032.0</td>\n",
       "      <td>0.042196</td>\n",
       "      <td>4366.0</td>\n",
       "      <td>550.0</td>\n",
       "      <td>0.009361</td>\n",
       "      <td>9.118037</td>\n",
       "      <td>31.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>45471.0</td>\n",
       "      <td>0.096017</td>\n",
       "      <td>13.265598</td>\n",
       "      <td>1.209562</td>\n",
       "      <td>0.068175</td>\n",
       "      <td>0.010996</td>\n",
       "      <td>0.057179</td>\n",
       "      <td>0.004398</td>\n",
       "      <td>9.601724</td>\n",
       "      <td>0.773947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Before_GAC_1</th>\n",
       "      <td>4472.0</td>\n",
       "      <td>3766.0</td>\n",
       "      <td>0.019573</td>\n",
       "      <td>2870.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>0.002843</td>\n",
       "      <td>5.523101</td>\n",
       "      <td>18.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>35063.0</td>\n",
       "      <td>0.081853</td>\n",
       "      <td>10.740667</td>\n",
       "      <td>0.593218</td>\n",
       "      <td>0.051336</td>\n",
       "      <td>0.025668</td>\n",
       "      <td>0.031372</td>\n",
       "      <td>0.005704</td>\n",
       "      <td>8.185267</td>\n",
       "      <td>0.479255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>After_GAC_1</th>\n",
       "      <td>5358.0</td>\n",
       "      <td>4264.0</td>\n",
       "      <td>0.017275</td>\n",
       "      <td>3561.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>0.003449</td>\n",
       "      <td>9.146341</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>46811.0</td>\n",
       "      <td>0.076072</td>\n",
       "      <td>9.108970</td>\n",
       "      <td>0.833138</td>\n",
       "      <td>0.006409</td>\n",
       "      <td>0.002136</td>\n",
       "      <td>0.004273</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.607186</td>\n",
       "      <td>0.414036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows  22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ARG_counts  scaffold_ARGs  scaffold_ARGs_normalized  \\\n",
       "HCL0_W2           6679.0         5248.0                  0.037242   \n",
       "HCL0_W3           7706.0         6032.0                  0.042196   \n",
       "Before_GAC_1      4472.0         3766.0                  0.019573   \n",
       "After_GAC_1       5358.0         4264.0                  0.017275   \n",
       "\n",
       "              scaffold_MGEs  scaffold_MGEs_with_ARGs  \\\n",
       "HCL0_W2              3831.0                    465.0   \n",
       "HCL0_W3              4366.0                    550.0   \n",
       "Before_GAC_1         2870.0                    208.0   \n",
       "After_GAC_1          3561.0                    390.0   \n",
       "\n",
       "              scaffold_MGEs_with_ARGs_normalized  \\\n",
       "HCL0_W2                                 0.008308   \n",
       "HCL0_W3                                 0.009361   \n",
       "Before_GAC_1                            0.002843   \n",
       "After_GAC_1                             0.003449   \n",
       "\n",
       "              %_scaffold_MGEs_with_ARGs/scaffold_ARGs  scaffold_pathogens  \\\n",
       "HCL0_W2                                      8.860518                31.0   \n",
       "HCL0_W3                                      9.118037                31.0   \n",
       "Before_GAC_1                                 5.523101                18.0   \n",
       "After_GAC_1                                  9.146341                 3.0   \n",
       "\n",
       "              scaffold_pathogens_with_ARGs  scaffold_pathogens_with_MGEs  ...  \\\n",
       "HCL0_W2                                7.0                          22.0  ...   \n",
       "HCL0_W3                                5.0                          26.0  ...   \n",
       "Before_GAC_1                           9.0                          11.0  ...   \n",
       "After_GAC_1                            1.0                           2.0  ...   \n",
       "\n",
       "              # scaffolds  scaffold_MGEs/#scaffolds  %scaffold_ARGs  \\\n",
       "HCL0_W2           40216.0                  0.095261       13.049533   \n",
       "HCL0_W3           45471.0                  0.096017       13.265598   \n",
       "Before_GAC_1      35063.0                  0.081853       10.740667   \n",
       "After_GAC_1       46811.0                  0.076072        9.108970   \n",
       "\n",
       "              %scaffold_MGEs_with_ARGs  %pathogens  %pathogens_with_ARGs  \\\n",
       "HCL0_W2                       1.156256    0.077084              0.017406   \n",
       "HCL0_W3                       1.209562    0.068175              0.010996   \n",
       "Before_GAC_1                  0.593218    0.051336              0.025668   \n",
       "After_GAC_1                   0.833138    0.006409              0.002136   \n",
       "\n",
       "              %pathogens_with_MGEs  %pathogens_with_ARGs_MGEs  %scaffold_MGEs  \\\n",
       "HCL0_W2                   0.054705                   0.004973        9.526059   \n",
       "HCL0_W3                   0.057179                   0.004398        9.601724   \n",
       "Before_GAC_1              0.031372                   0.005704        8.185267   \n",
       "After_GAC_1               0.004273                   0.000000        7.607186   \n",
       "\n",
       "              mean_cov_MGEs_with_ARGs/mean_cov  \n",
       "HCL0_W2                               0.718541  \n",
       "HCL0_W3                               0.773947  \n",
       "Before_GAC_1                          0.479255  \n",
       "After_GAC_1                           0.414036  \n",
       "\n",
       "[4 rows x 22 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MGEs_and_ARGs_pathogens_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "runs 'ARG_subtype_count'\n",
    "-----------------------\n",
    "\n",
    "\n",
    "Run 'scaffold_count' before with desired scaffold length threshhold.\n",
    "\n",
    "Returns a dataframe with count of ARG subtype.\n",
    "\"\"\"\n",
    "\n",
    "columns = ['arg_subtype_count', 'scaffold_arg_subtype', \n",
    "           'scaffold_has_arg_subtype_normalized']\n",
    "ARG_subtypes = pd.DataFrame(columns=columns, dtype=float)\n",
    "for name, df in all_full_tables_assembly_data.items():    \n",
    "    count = ARG_subtype_count(df, min_bitscore=50, \n",
    "                      min_id_arg=50, min_coverage_arg = 0.5, arg_name = 'msrE')\n",
    "    count_series = pd.Series(count)\n",
    "    count_rename = count_series.rename(name)\n",
    "    \n",
    "    ARG_subtypes = ARG_subtypes.append(count_rename)\n",
    "        \n",
    "\n",
    "ARG_subtypes_store = ARG_subtypes \n",
    "\n",
    "\n",
    "ARG_subtypes_scaffold_count = \\\n",
    "pd.concat([ARG_subtypes_store, scaffold_count], axis=1, join='inner')\n",
    "\n",
    "ARG_subtypes_scaffold_count['%scaffold_arg_subtype']=\\\n",
    "ARG_subtypes_scaffold_count['scaffold_arg_subtype']*100/\\\n",
    "ARG_subtypes_scaffold_count['# scaffolds']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>arg_subtype_count</th>\n",
       "      <th>scaffold_arg_subtype</th>\n",
       "      <th>scaffold_has_arg_subtype_normalized</th>\n",
       "      <th># scaffolds</th>\n",
       "      <th>%scaffold_arg_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCL0_W2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>40216.0</td>\n",
       "      <td>0.002487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>45471.0</td>\n",
       "      <td>0.002199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Before_GAC_1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>35063.0</td>\n",
       "      <td>0.002852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>After_GAC_1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>46811.0</td>\n",
       "      <td>0.002136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              arg_subtype_count  scaffold_arg_subtype  \\\n",
       "HCL0_W2                     1.0                   1.0   \n",
       "HCL0_W3                     1.0                   1.0   \n",
       "Before_GAC_1                1.0                   1.0   \n",
       "After_GAC_1                 1.0                   1.0   \n",
       "\n",
       "              scaffold_has_arg_subtype_normalized  # scaffolds  \\\n",
       "HCL0_W2                                  0.000010      40216.0   \n",
       "HCL0_W3                                  0.000019      45471.0   \n",
       "Before_GAC_1                             0.000005      35063.0   \n",
       "After_GAC_1                              0.000003      46811.0   \n",
       "\n",
       "              %scaffold_arg_subtype  \n",
       "HCL0_W2                    0.002487  \n",
       "HCL0_W3                    0.002199  \n",
       "Before_GAC_1               0.002852  \n",
       "After_GAC_1                0.002136  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARG_subtypes_scaffold_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Step necessary before merging samples\n",
    "-------------------------------------\n",
    "\n",
    "Adds sample names (exp. 'HCL0_W2') to \n",
    "scaffold numbers in index (exp. scaffold_0).\n",
    "The modified index will have both sample name \n",
    "and scaffold number.\n",
    "Dataframes will be updated with the new combined index\n",
    "and stored in the same dictionary called\n",
    "all_full_tables_remove_overlap.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "for name, df in all_full_tables_remove_overlap.items():\n",
    "    df['file_name'] = name\n",
    "    df['index'] = df.index\n",
    "    df['read'] = df['file_name'] + df['index']\n",
    "    df.index = df['read']\n",
    "    df = df.drop(['read', 'file_name', 'index'], axis=1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "MERGING SAMPLES TO DESIRED GROUPS\n",
    "---------------------------------\n",
    "\n",
    "\n",
    "Combine samples to be analyzed as a group\n",
    "and store as pd.DataFrame.\n",
    "\n",
    "For example,\n",
    "HCL0_W2 and HCL0_W3 are two NanoARG tables, \n",
    "chlorine_unfiltered is the combined table.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "chlorine_unfiltered = pd.concat([\n",
    "                                all_full_tables_remove_overlap['HCL0_W2'],\n",
    "                                all_full_tables_remove_overlap['HCL0_W3']\n",
    "                                ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100.00%    \r"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Running 'ARG_MGE_association'\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "Change 'chlorine_unfiltered' to your sample name.\n",
    "Change parameters as desired. Parameter not specified runs with default.\n",
    "\n",
    "Set the ARG aa idendity higher and lower, change max_fdr etc. to explore the data.\n",
    "For example, setting high identity and coverage can return ARGs \n",
    "known to be associated with mobile genetic elements.\n",
    "\n",
    "Import/deduce 'f' from dataframe 'MGEs_and_ARGs_pathogens_normalized'. \n",
    "\n",
    "Group samples differently for data exploration. \n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "chlorine_unfiltered_ARG_MGE_association = ARG_MGE_association(\n",
    "    chlorine_unfiltered, min_id_arg=50, \n",
    "    f = 0.01,\n",
    "    min_coverage_arg=0.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Running 'neighbor_stretch'\n",
    "--------------------------\n",
    "Change 'chlorine_unfiltered' to your sample name.\n",
    "Change parameters as desired. \n",
    "For example, setting max_distance higher looks at a longer stretch of the scaffold.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "neighbor_stretch_unfiltered_chlorine = neighbor_stretch(\n",
    "    chlorine_unfiltered, min_ARG_count=1, max_distance =5000, \n",
    "                     number_neighboring_orf=4, \n",
    "                     sliding_window_step_increase=4, min_bitscore=50, \n",
    "                     min_id_arg=80, min_id_mge=25, min_coverage_arg=0.5, verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>group</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>identity</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>evalue</th>\n",
       "      <th>NCBI_taxa_id</th>\n",
       "      <th>taxa_centrifuge_score</th>\n",
       "      <th>species</th>\n",
       "      <th>coverage</th>\n",
       "      <th>is_pathogen</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCL0_W2scaffold_3375</th>\n",
       "      <td>ANP63073.1</td>\n",
       "      <td>mphD</td>\n",
       "      <td>ARGs</td>\n",
       "      <td>MLS</td>\n",
       "      <td>4070</td>\n",
       "      <td>4882</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "      <td>548.1</td>\n",
       "      <td>1.600000e-155</td>\n",
       "      <td>48296</td>\n",
       "      <td>302771.0</td>\n",
       "      <td>Acinetobacter pittii</td>\n",
       "      <td>0.921769</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2scaffold_3375</th>\n",
       "      <td>YP_724476.1</td>\n",
       "      <td>msrE</td>\n",
       "      <td>ARGs</td>\n",
       "      <td>MLS</td>\n",
       "      <td>4941</td>\n",
       "      <td>6413</td>\n",
       "      <td>+</td>\n",
       "      <td>100.0</td>\n",
       "      <td>943.0</td>\n",
       "      <td>2.300000e-274</td>\n",
       "      <td>48296</td>\n",
       "      <td>302771.0</td>\n",
       "      <td>Acinetobacter pittii</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2scaffold_3375</th>\n",
       "      <td>UniRef90_P55373</td>\n",
       "      <td>Cluster: Putative transposase y4bF</td>\n",
       "      <td>Functional Genes</td>\n",
       "      <td>Cluster: Putative transposase y4bF</td>\n",
       "      <td>6968</td>\n",
       "      <td>8221</td>\n",
       "      <td>+</td>\n",
       "      <td>46.9</td>\n",
       "      <td>376.3</td>\n",
       "      <td>9.200000e-103</td>\n",
       "      <td>48296</td>\n",
       "      <td>302771.0</td>\n",
       "      <td>Acinetobacter pittii</td>\n",
       "      <td>0.923414</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2scaffold_3375</th>\n",
       "      <td>WP_067143567.1</td>\n",
       "      <td>transposase [Oceanivirga salmonicida]</td>\n",
       "      <td>MGEs</td>\n",
       "      <td>transposase</td>\n",
       "      <td>7331</td>\n",
       "      <td>8017</td>\n",
       "      <td>+</td>\n",
       "      <td>30.9</td>\n",
       "      <td>100.1</td>\n",
       "      <td>1.800000e-19</td>\n",
       "      <td>48296</td>\n",
       "      <td>302771.0</td>\n",
       "      <td>Acinetobacter pittii</td>\n",
       "      <td>0.682493</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_id                              gene_name  \\\n",
       "read                                                                           \n",
       "HCL0_W2scaffold_3375       ANP63073.1                                   mphD   \n",
       "HCL0_W2scaffold_3375      YP_724476.1                                   msrE   \n",
       "HCL0_W2scaffold_3375  UniRef90_P55373     Cluster: Putative transposase y4bF   \n",
       "HCL0_W2scaffold_3375   WP_067143567.1  transposase [Oceanivirga salmonicida]   \n",
       "\n",
       "                                 group                            category  \\\n",
       "read                                                                         \n",
       "HCL0_W2scaffold_3375              ARGs                                 MLS   \n",
       "HCL0_W2scaffold_3375              ARGs                                 MLS   \n",
       "HCL0_W2scaffold_3375  Functional Genes  Cluster: Putative transposase y4bF   \n",
       "HCL0_W2scaffold_3375              MGEs                         transposase   \n",
       "\n",
       "                      start   end strand  identity  bitscore         evalue  \\\n",
       "read                                                                          \n",
       "HCL0_W2scaffold_3375   4070  4882      +     100.0     548.1  1.600000e-155   \n",
       "HCL0_W2scaffold_3375   4941  6413      +     100.0     943.0  2.300000e-274   \n",
       "HCL0_W2scaffold_3375   6968  8221      +      46.9     376.3  9.200000e-103   \n",
       "HCL0_W2scaffold_3375   7331  8017      +      30.9     100.1   1.800000e-19   \n",
       "\n",
       "                     NCBI_taxa_id  taxa_centrifuge_score  \\\n",
       "read                                                       \n",
       "HCL0_W2scaffold_3375        48296               302771.0   \n",
       "HCL0_W2scaffold_3375        48296               302771.0   \n",
       "HCL0_W2scaffold_3375        48296               302771.0   \n",
       "HCL0_W2scaffold_3375        48296               302771.0   \n",
       "\n",
       "                                   species  coverage  is_pathogen  \n",
       "read                                                               \n",
       "HCL0_W2scaffold_3375  Acinetobacter pittii  0.921769            0  \n",
       "HCL0_W2scaffold_3375  Acinetobacter pittii  1.000000            0  \n",
       "HCL0_W2scaffold_3375  Acinetobacter pittii  0.923414            0  \n",
       "HCL0_W2scaffold_3375  Acinetobacter pittii  0.682493            0  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neighbor_stretch_unfiltered_chlorine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
