{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from os.path import join\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "from scipy.stats import binom_test\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nImports and merges NanoARG Full Table and assembly information (length and read-count).\\nCalculates quantities and relative abundances of scaffolds with ARGs, MRGs, MGEs and \\npathogen scaffols. \\n\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Imports and merges NanoARG Full Table and assembly information (length and read-count).\n",
    "Calculates quantities and relative abundances of scaffolds with ARGs, MRGs, MGEs and \n",
    "pathogen scaffols. \n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_overlap_not_ARG_strand_selection(df, \n",
    "                                            max_overlap=250, \n",
    "                                            verbose=False, \n",
    "                                            strand_selection='no'):\n",
    "    \"\"\" \n",
    "    Rearrange open reading frames according to position.\n",
    "    Find open reading frames identified as ARGs that overlap with other \n",
    "    (MGEs, MRGs, Functional Genes) open reading frames and \n",
    "    remove them. \n",
    "    It is necessary to run this before many of the other functions.\n",
    "    For example, the same region identified as ARG and MGE can result in \n",
    "    misidentification of contigs with both ARG and MGE. Overlapping \n",
    "    open reading frame identifications also make it more difficult to analyze \n",
    "    contiguous regions.\n",
    "    \n",
    "    Example\n",
    "    -------\n",
    "          \n",
    "    s1 ----ARGs------- e1 \n",
    "         s2 --------- e2 removed if overlap > max_overlap\n",
    "                  s3-------------e3 keep if overlap < max_overlap\n",
    "                           s4 ------ e4 \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        data frame with NanoARG table\n",
    "    max_overlap : int\n",
    "        maximum allowed overlap between consequitive orfs\n",
    "    \n",
    "    strand_selection : string\n",
    "        choose 'no','yes'\n",
    "        \n",
    "        default run with 'no', preferred for subsequent data analysis\n",
    "        ignore +/- strand\n",
    "        (11) scaffold_10 4879-6615 ARGs\n",
    "           (12) 5947-6558  removing Functional Genes! overlap 611\n",
    "        (27) scaffold_10 44937-45410 ARGs\n",
    "        (30) scaffold_10 47395-49008 ARGs\n",
    "           (31) 47395-48969  removing MRGs! overlap 1574\n",
    "           (32) 47443-48996  removing Functional Genes! overlap 1553\n",
    "        (36) scaffold_10 52079-52978 ARGs\n",
    "        (37) scaffold_10 56260-57171 ARGs\n",
    "           (38) 56260-57168  removing Functional Genes! overlap 908\n",
    "        (42) scaffold_10002 34-753 ARGs\n",
    "           (43) 40-768  removing Functional Genes! overlap 713\n",
    "           \n",
    "        run with 'yes'\n",
    "        only remove if orfs have same 5'-> 3' direction\n",
    "        \n",
    "        (11) scaffold_10 4879-6615 ARGs\n",
    "           (12) 5947-6558 strand + + removing Functional Genes! overlap 611\n",
    "        (27) scaffold_10 44937-45410 ARGs\n",
    "        (30) scaffold_10 47395-49008 ARGs\n",
    "           (31) 47395-48969 strand + - overlap 1574\n",
    "           (32) 47443-48996 strand + - overlap 1553\n",
    "        (36) scaffold_10 52079-52978 ARGs\n",
    "        (37) scaffold_10 56260-57171 ARGs\n",
    "           (38) 56260-57168 strand + - overlap 908\n",
    "        (42) scaffold_10002 34-753 ARGs\n",
    "           (43) 40-768 strand + + removing Functional Genes! overlap 713\n",
    "        \n",
    "\n",
    "    verbose : bool\n",
    "        print debug info\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    Dataframe with entries removed\n",
    "    \n",
    "    \n",
    "   \n",
    "    \"\"\"\n",
    "    if verbose: \n",
    "        print(\"sorting...\")\n",
    "        \n",
    "    group_column = 2\n",
    "\n",
    "    D_sorted = df.groupby(df.index).apply(\n",
    "        lambda x: x.sort_values('start', ascending=True))\n",
    "    \n",
    "    Ddrop = D_sorted.droplevel(level=0).copy()\n",
    "\n",
    "    if verbose: \n",
    "        print(\"searching\")\n",
    "\n",
    "    for i in range(len(Ddrop)-1):\n",
    "        if Ddrop.group.iloc[i] == 'ARGs':\n",
    "            scaffold = Ddrop.index[i]\n",
    "\n",
    "            if verbose: \n",
    "                print(\"({}) {} {}-{} {}\".format(\n",
    "                    i, scaffold, Ddrop.start.iloc[i], Ddrop.end.iloc[i], Ddrop.group.iloc[i]))\n",
    "\n",
    "            j = i + 1\n",
    "            while (j<len(Ddrop) and (Ddrop.start.iloc[j] <= Ddrop.end.iloc[i])\n",
    "                   and (Ddrop.index[j] == scaffold)):\n",
    "\n",
    "                if verbose:\n",
    "                    print(\"   ({}) {}-{} \".format(\n",
    "                        j, Ddrop.start.iloc[j], Ddrop.end.iloc[j]), end='')\n",
    "\n",
    "                overlap = np.min([Ddrop.end.iloc[j], Ddrop.end.iloc[i]]) - Ddrop.start.iloc[j]\n",
    "                if overlap >= max_overlap:\n",
    "                    if Ddrop.group.iloc[j] != 'ARGs':\n",
    "                        if strand_selection == 'yes':\n",
    "                            if verbose:\n",
    "                                print(\"strand {} {}\".format(\n",
    "                                    Ddrop.strand.iloc[i], Ddrop.strand.iloc[j]), end = '')\n",
    "                               \n",
    "                            if Ddrop.strand.iloc[j] == Ddrop.strand.iloc[i]:\n",
    "                                if verbose:\n",
    "                                    print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                                Ddrop.iat[j, group_column] = np.nan\n",
    "                        elif strand_selection == 'no':\n",
    "                            if verbose:\n",
    "                                print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                            Ddrop.iat[j, group_column] = np.nan\n",
    "                else:\n",
    "\n",
    "                    if verbose:\n",
    "                        print(\"keeping {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "\n",
    "                if verbose:\n",
    "                    print(\" overlap {}\".format(overlap))\n",
    "\n",
    "                j = j + 1      \n",
    "                \n",
    "    for i in range(1, len(Ddrop)):\n",
    "        if Ddrop.group.iloc[i] == 'ARGs':\n",
    "            scaffold = Ddrop.index[i]\n",
    "\n",
    "            if verbose: \n",
    "                print(\"({}) {} {}-{} {}\".format(\n",
    "                    i, scaffold, Ddrop.start.iloc[i], Ddrop.end.iloc[i], Ddrop.group.iloc[i]))\n",
    "\n",
    "     \n",
    "            j = i - 1\n",
    "            while (j>=0 and (Ddrop.start.iloc[i] <= Ddrop.end.iloc[j])\n",
    "                   and (Ddrop.index[j] == scaffold)):\n",
    "\n",
    "                if verbose:\n",
    "                    print(\"   ({}) {}-{} \".format(\n",
    "                        j, Ddrop.start.iloc[j], Ddrop.end.iloc[j]), end='')\n",
    "\n",
    "                overlap = np.min([Ddrop.end.iloc[j], Ddrop.end.iloc[i]]) - Ddrop.start.iloc[i]\n",
    "                if overlap >= max_overlap:\n",
    "                    if Ddrop.group.iloc[j] != 'ARGs':\n",
    "                        if strand_selection == 'yes':\n",
    "                            if verbose:\n",
    "                                print(\"strand {} {}\".format(\n",
    "                                    Ddrop.strand.iloc[i], Ddrop.strand.iloc[j]), end='')\n",
    "                            if Ddrop.strand.iloc[j] == Ddrop.strand.iloc[i]:\n",
    "                                if verbose:\n",
    "                                    print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                                Ddrop.iat[j, group_column] = np.nan\n",
    "                        elif strand_selection == 'no':\n",
    "                            if verbose:\n",
    "                                print(\" removing {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "                            Ddrop.iat[j, group_column] = np.nan\n",
    "                    \n",
    "                else:\n",
    "\n",
    "                    if verbose:\n",
    "                        print(\"keeping {}!\".format(Ddrop.group.iloc[j]), end='')\n",
    "\n",
    "                if verbose:\n",
    "                    print(\" overlap {}\".format(overlap))\n",
    "\n",
    "                j = j - 1\n",
    "                \n",
    "    df_remove_overlap = Ddrop.dropna(subset=['group']) \n",
    "    \n",
    "    return df_remove_overlap       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Necessary to run 'remove_overlap_not_ARG_strand_selection' before. \n",
    "Calculates quantities and relative abundances of scaffolds with ARGs,\n",
    "MGEs and scaffolds identified as pathogens.\n",
    "\"\"\"\n",
    "\n",
    "def count_scaffolds_with_MGEs_and_ARGs_normalized(df, min_score=40000, \n",
    "                                                  min_length = 500):\n",
    "    \"\"\"\n",
    "    Count\n",
    "        ARG_counts\n",
    "        scaffold_ARGs\n",
    "        scaffold_ARGs_normalized\n",
    "        scaffold_MGEs\n",
    "        scaffold_MGEs_with_ARGs\n",
    "        scaffold_MGEs_with_ARGs_normalized\n",
    "        %_scaffold_MGEs_with_ARGs/scaffold_ARGs\n",
    "        scaffold_pathogens\n",
    "        scaffold_pathogens_with_args\n",
    "        scaffold_pathogens_with_mges\n",
    "        scaffold_pathogens_with_args_mges\n",
    "        scaffold_has_pathogens_ARGs_MGEs_normalized\n",
    "    \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold, ARGs and MGEs data\n",
    "    min_score : int\n",
    "        taxa_centrifuge_score cutoff for taxonomy classification\n",
    "    min_length : int\n",
    "        minimum length of scaffold to include in analysis\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dict:\n",
    "        dictionary with ratio and counts\n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df_length_filtered = df[df['length'] > min_length]\n",
    "    \n",
    "    scaffold_has_args_normalized = 0\n",
    "    scaffold_has_args=0\n",
    "    old_index_arg = 0\n",
    "    for i in range(len(df_length_filtered)):\n",
    "        if df_length_filtered.group.iloc[i] == 'ARGs':\n",
    "            if df_length_filtered.index[i] != old_index_arg:\n",
    "                scaffold_has_args += 1\n",
    "                scaffold_has_args_normalized  += (\n",
    "                    df_length_filtered.normalized_read_count_per_kb.iloc[i])\n",
    "                old_index_arg = df_length_filtered.index[i]\n",
    "    \n",
    "    \n",
    "    scaffold_has_mges = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: (x['group']=='MGEs').any()\n",
    "    ).sum()\n",
    "    \n",
    "    scaffold_has_mges_with_args=0\n",
    "    scaffold_has_mges_with_args_norm=0\n",
    "    same_index=0\n",
    "    for scaffold, data in df_length_filtered.groupby(level=0):\n",
    "        groups = np.array(data.group)\n",
    "        normalized_read_count_per_kbs = np.array(data.normalized_read_count_per_kb)\n",
    "        indices = np.array(data.index)\n",
    "        for i in range(len(groups)):\n",
    "            if (groups[i] == 'ARGs'):\n",
    "                for j in range(len(groups)):\n",
    "                    if (groups[j] == 'MGEs'):\n",
    "                        if indices[j] != same_index:\n",
    "                            scaffold_has_mges_with_args += 1\n",
    "                            scaffold_has_mges_with_args_norm += normalized_read_count_per_kbs[j]\n",
    "                            same_index = indices[j]\n",
    "        \n",
    "    \n",
    "    scaffold_has_pathogens = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: \n",
    "            (x['is_pathogen']==1).any() \n",
    "            and (x['taxa_centrifuge_score']>min_score).any()\n",
    "    ).sum()\n",
    "    \n",
    "    scaffold_has_pathogens_args = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: \n",
    "            (x['is_pathogen']==1).any() \n",
    "            and (x['taxa_centrifuge_score']>min_score).any() \n",
    "            and (x['group']=='ARGs').any()\n",
    "    ).sum()\n",
    "    \n",
    "    scaffold_has_pathogens_mges = df_length_filtered.groupby(level=0).apply(\n",
    "        lambda x: \n",
    "            (x['is_pathogen']==1).any() \n",
    "            and (x['taxa_centrifuge_score']>min_score).any()\n",
    "            and (x['group']=='MGEs').any() \n",
    "    ).sum()\n",
    "    \n",
    "    df_length_taxa_filtered = df_length_filtered[df_length_filtered['taxa_centrifuge_score'] > min_score]\n",
    "    scaffold_has_pathogens_args_mges=0\n",
    "    scaffold_has_pathogens_args_mges_norm=0\n",
    "    same_index_pathogen=0\n",
    "    for scaffold, data in df_length_taxa_filtered.groupby(level=0):\n",
    "        groups = np.array(data.group)\n",
    "        normalized_read_count_per_kbs = np.array(data.normalized_read_count_per_kb)\n",
    "        indices = np.array(data.index)\n",
    "        pathogens = np.array(data.is_pathogen)\n",
    "        for i in range(len(groups)):\n",
    "            if (pathogens[i] == 1):\n",
    "                for j in range(len(groups)):\n",
    "                    if (groups[j] == 'ARGs'):\n",
    "                        for k in range(len(groups)):\n",
    "                            if (groups[k] == 'MGEs'):\n",
    "                                if indices[k] != same_index_pathogen:\n",
    "                                    scaffold_has_pathogens_args_mges += 1\n",
    "                                    scaffold_has_pathogens_args_mges_norm += normalized_read_count_per_kbs[k]\n",
    "                                    same_index_pathogen = indices[k]\n",
    "    \n",
    "    ARG_counts = len(df_length_filtered[df_length_filtered['group'] == 'ARGs'])\n",
    "    \n",
    "    \n",
    "    return {\n",
    "        'ARG_counts' : ARG_counts,\n",
    "        'scaffold_ARGs': scaffold_has_args,\n",
    "        'scaffold_ARGs_normalized' : scaffold_has_args_normalized,\n",
    "        'scaffold_MGEs': scaffold_has_mges,\n",
    "        'scaffold_MGEs_with_ARGs': scaffold_has_mges_with_args,\n",
    "        'scaffold_MGEs_with_ARGs_normalized' :scaffold_has_mges_with_args_norm,\n",
    "        '%_scaffold_MGEs_with_ARGs/scaffold_ARGs': (\n",
    "            scaffold_has_mges_with_args / scaffold_has_args * 100\n",
    "        ),\n",
    "        'scaffold_pathogens' : scaffold_has_pathogens,\n",
    "        'scaffold_pathogens_with_ARGs' : scaffold_has_pathogens_args,\n",
    "        'scaffold_pathogens_with_MGEs' : scaffold_has_pathogens_mges,\n",
    "        'scaffold_pathogens_with_ARGs_MGEs' : scaffold_has_pathogens_args_mges,\n",
    "        'scaffold_has_pathogens_ARGs_MGEs_normalized' : scaffold_has_pathogens_args_mges_norm\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Returns quantity and relative abundance of desired ARG subtype scaffold. \n",
    "\"\"\"\n",
    "\n",
    "def ARG_subtype_count(df, min_bitscore=50, \n",
    "                      min_id_arg=25, min_coverage_arg=0.4, arg_name = 'sul1'):\n",
    "    \n",
    "    \"\"\"\n",
    "    Count\n",
    "        arg_subtype_count = count # of ARG subtype occurance\n",
    "        scaffold_arg_subtype = count # of scaffolds with ARG_subtype\n",
    "        scaffold_has_arg_subtype_normalized = scaffolds with ARG_subtype in RPKM\n",
    "    \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold, ARGs and MGEs data\n",
    "    min_bitscore : int\n",
    "        minimum bitscore, default is 50\n",
    "    min_id_arg : int\n",
    "        minimum identity of ARG, default is 25    \n",
    "    min_coverage_arg : int\n",
    "        minimum coverage of ARG, default is 0.4 in NanoARG\n",
    "    arg_name : name of ARG subtype to be counted\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dict:\n",
    "        dictionary with counts   \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    \n",
    "    df_bitscore_adjusted = df[df['bitscore'] > min_bitscore]\n",
    "    \n",
    "    df_bit_arg_adjusted = df_bitscore_adjusted.drop(df_bitscore_adjusted[ \n",
    "        (df_bitscore_adjusted['group'] == 'ARGs') \n",
    "        & ((df_bitscore_adjusted['identity'] < min_id_arg)\n",
    "            | (df_bitscore_adjusted['coverage'] < min_coverage_arg))].index)\n",
    "    \n",
    "    \n",
    "    arg_subtype_count = len(df_bit_arg_adjusted[df_bit_arg_adjusted['gene_name'] == arg_name])\n",
    "    \n",
    "    scaffold_has_arg_subtype = df_bit_arg_adjusted.groupby(level=0).apply(\n",
    "        lambda x: (x['gene_name']==arg_name).any()\n",
    "    ).sum()\n",
    "    \n",
    "    \n",
    "    scaffold_has_arg_subtype_normalized = 0\n",
    "    old_index = 0\n",
    "    for i in range(len(df_bit_arg_adjusted)):\n",
    "        if df_bit_arg_adjusted.gene_name.iloc[i] == arg_name:\n",
    "            if df_bit_arg_adjusted.index[i] != old_index:              \n",
    "                scaffold_has_arg_subtype_normalized  += (\n",
    "                    df_bit_arg_adjusted.normalized_read_count_per_kb.iloc[i])\n",
    "                old_index = df_bit_arg_adjusted.index[i]\n",
    "                \n",
    "    return {\n",
    "        'arg_subtype_count' : arg_subtype_count,\n",
    "        'scaffold_arg_subtype' : scaffold_has_arg_subtype,\n",
    "        'scaffold_has_arg_subtype_normalized' : scaffold_has_arg_subtype_normalized\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_scaffolds(df, min_length=500):\n",
    "    \"\"\"    \n",
    "    Count number of scaffolds longer than min_length \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.dataframe\n",
    "        dataframe with scaffold length and read-count\n",
    "    min_length : int\n",
    "        minimum length of scaffold to include in analysis\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    dict:\n",
    "        dictionary with counts\n",
    "           \n",
    "    \"\"\"\n",
    "    \n",
    "    df_length_filtered = df[df['length'] > min_length]\n",
    "    scaffold_counts = len(df_length_filtered.groupby(level=0))\n",
    "    \n",
    "    return {\n",
    "        '# scaffolds' : scaffold_counts,\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "INPUT REQUIRED. IMPORTING NANOARG DATA.\n",
    "------------------------------------------------------\n",
    "\n",
    "Import multiple NanoARG Full Tables (.tsv) and save in a dictionary.\n",
    "\n",
    "Input required:\n",
    "1. Path to the directory with NanoARG full tables\n",
    "\n",
    "Output:\n",
    "1. Disctionary called all_full_tables, each sample is saved \n",
    "    in a different dataframe with sample name.\n",
    " \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "***provide PATH to the directory with files*********\n",
    "\n",
    "\"\"\"\n",
    "path = '/Users/sudeshnaghosh/Dropbox/VT/test2/'\n",
    "\n",
    "\n",
    "#------------------------------------------------\n",
    "\n",
    "\n",
    "DATA_DIR = path\n",
    "\n",
    "all_full_tables = {}\n",
    "\n",
    "for file in os.listdir(DATA_DIR):\n",
    "    df_name = file.split('.')[0]\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    all_full_tables[df_name] = pd.read_csv(DATA_DIR + file, sep=',', header=0, index_col=0)\n",
    "    \n",
    "del all_full_tables[\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/After_GAC_1.fa\n",
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/Before_GAC_1.fa\n",
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/HCL0_W2.fa\n",
      "/Users/sudeshnaghosh/Dropbox/VT/assembly_test/HCL0_W3.fa\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "INPUT REQUIRED. IMPORTING ASSEMBLY INFORMATION.\n",
    "------------------------------------------------------\n",
    "\n",
    "Extract length and read_count information about each scaffold \n",
    "in a sample and store as dataframe. Store dataframes in a dictionary.\n",
    "\n",
    "\n",
    "Input:\n",
    "\n",
    "Input file format, from IDBA-UD\n",
    "\n",
    ">scaffold_0 length_285165 read_count_39166\n",
    "CGGGCCCTACCGTAGCAGCCGCTACGGTAGGGCCCGTGTCCAGT......\n",
    ">scaffold_1 length_297291 read_count_43258\n",
    "GAGAGCGTCATCGATCATCGCGACGCAGGTTAACAGCGATTGCCGATGTTTACAACC........\n",
    "\n",
    "Output:\n",
    "    \n",
    "Dictionary called assembly\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "***provide PATH to the directory with IDBA-UD files**********\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "path = '/Users/sudeshnaghosh/Dropbox/VT/assembly_test'\n",
    "#---------------------------------------------------\n",
    "\n",
    "\n",
    "DATA_DIR = path\n",
    "\n",
    "assembly = {}\n",
    "\n",
    "\n",
    "for file in os.listdir(DATA_DIR):\n",
    "    if file.endswith(\".fa\"):\n",
    "        file_path = join(DATA_DIR, file)\n",
    "        print(file_path)\n",
    "        df_name = file.split('.')[0]\n",
    "        scaffold = []\n",
    "        length = []\n",
    "        read_count = []\n",
    "        \n",
    "        for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "            scaffold.append(record.description.split()[0])\n",
    "            length.append(record.description.split()[1].split('_')[1])\n",
    "            read_count.append(record.description.rsplit()[2].split('_')[2])\n",
    "            \n",
    "            \n",
    "        assembly[df_name] = pd.DataFrame({\n",
    "            'read':scaffold,\n",
    "            'length':length,\n",
    "            'read_count':read_count\n",
    "        })\n",
    "        assembly[df_name].set_index('read', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "For each scaffold, calculate rpm and rpkm \n",
    "-----------------------------------------\n",
    "\n",
    "\n",
    "For each scaffold, calculate rpm and rpkm and add to the disctionary with \n",
    "length and read_count.\n",
    "\n",
    "normalized_read_count (rpm) = \n",
    "read_count of scaffold / (total # of read_counts in a sample / 1,000,000)\n",
    "normalized_read_count_per_kb (rpkm) = \n",
    "normalized_read_count (rpm) / length of scaffold in kb\n",
    "\n",
    "The updated dataframes are stored in the dictionary called assembly.\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "for name, keys in assembly.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    SUM = ((assembly[df_name]['read_count']).astype(float)).sum()\n",
    "    assembly[df_name] = assembly[df_name].astype(float)\n",
    "    assembly[df_name]['normalized_read_count'] = (\n",
    "        (assembly[df_name]['read_count'] * 1000000) / SUM)\n",
    "    assembly[df_name]['normalized_read_count_per_kb'] = (\n",
    "        (assembly[df_name]['normalized_read_count'] * 1000)/\n",
    "        assembly[df_name]['length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "      <th>normalized_read_count</th>\n",
       "      <th>normalized_read_count_per_kb</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_0</th>\n",
       "      <td>709020.0</td>\n",
       "      <td>97243.0</td>\n",
       "      <td>6486.089388</td>\n",
       "      <td>9.147964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_1</th>\n",
       "      <td>551495.0</td>\n",
       "      <td>312878.0</td>\n",
       "      <td>20868.902395</td>\n",
       "      <td>37.840601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_2</th>\n",
       "      <td>954535.0</td>\n",
       "      <td>669486.0</td>\n",
       "      <td>44654.587376</td>\n",
       "      <td>46.781509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_3</th>\n",
       "      <td>303743.0</td>\n",
       "      <td>175890.0</td>\n",
       "      <td>11731.829155</td>\n",
       "      <td>38.624196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_4</th>\n",
       "      <td>302455.0</td>\n",
       "      <td>32823.0</td>\n",
       "      <td>2189.287784</td>\n",
       "      <td>7.238392</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              length  read_count  normalized_read_count  \\\n",
       "read                                                      \n",
       "scaffold_0  709020.0     97243.0            6486.089388   \n",
       "scaffold_1  551495.0    312878.0           20868.902395   \n",
       "scaffold_2  954535.0    669486.0           44654.587376   \n",
       "scaffold_3  303743.0    175890.0           11731.829155   \n",
       "scaffold_4  302455.0     32823.0            2189.287784   \n",
       "\n",
       "            normalized_read_count_per_kb  \n",
       "read                                      \n",
       "scaffold_0                      9.147964  \n",
       "scaffold_1                     37.840601  \n",
       "scaffold_2                     46.781509  \n",
       "scaffold_3                     38.624196  \n",
       "scaffold_4                      7.238392  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly['After_GAC_1'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "COMBINING NANOARG FULL TABLES AND ASSEMBLY INFORMATION\n",
    "------------------------------------------------------\n",
    "\n",
    "Add scaffold length, read_count and normalized scaffold read_count \n",
    "(rpm and rpkm) to NanoARG table.\n",
    "Store dataframes in a dictionary called all_full_tables_assembly_data.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "all_full_tables_assembly_data = {}\n",
    "\n",
    "for name, keys in all_full_tables.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    all_full_tables_assembly_data[df_name] = (\n",
    "        pd.concat([all_full_tables[df_name], assembly[df_name]], \n",
    "                  axis=1, sort=False, join='inner'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>group</th>\n",
       "      <th>category</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>identity</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>evalue</th>\n",
       "      <th>NCBI_taxa_id</th>\n",
       "      <th>taxa_centrifuge_score</th>\n",
       "      <th>species</th>\n",
       "      <th>coverage</th>\n",
       "      <th>is_pathogen</th>\n",
       "      <th>length</th>\n",
       "      <th>read_count</th>\n",
       "      <th>normalized_read_count</th>\n",
       "      <th>normalized_read_count_per_kb</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>scaffold_16508</th>\n",
       "      <td>WP_007306505.1</td>\n",
       "      <td>IS1634 family transposase [Crocosphaera watsonii]</td>\n",
       "      <td>MGEs</td>\n",
       "      <td>transposase</td>\n",
       "      <td>1058</td>\n",
       "      <td>1459</td>\n",
       "      <td>+</td>\n",
       "      <td>35.1</td>\n",
       "      <td>102.4</td>\n",
       "      <td>6.500000e-21</td>\n",
       "      <td>undefined</td>\n",
       "      <td>0.0</td>\n",
       "      <td>undefined</td>\n",
       "      <td>0.724324</td>\n",
       "      <td>0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>4.935786</td>\n",
       "      <td>2.831776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_9236</th>\n",
       "      <td>YP_002467958</td>\n",
       "      <td>PBP-1B</td>\n",
       "      <td>ARGs</td>\n",
       "      <td>beta-lactam</td>\n",
       "      <td>326</td>\n",
       "      <td>1816</td>\n",
       "      <td>+</td>\n",
       "      <td>26.6</td>\n",
       "      <td>137.9</td>\n",
       "      <td>1.300000e-32</td>\n",
       "      <td>undefined</td>\n",
       "      <td>0.0</td>\n",
       "      <td>undefined</td>\n",
       "      <td>0.718421</td>\n",
       "      <td>0</td>\n",
       "      <td>2392.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>5.402684</td>\n",
       "      <td>2.258647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_9236</th>\n",
       "      <td>UniRef90_Q07259</td>\n",
       "      <td>Cluster: Putative transglycosylase H16 A0665</td>\n",
       "      <td>Functional Genes</td>\n",
       "      <td>Cluster: Putative transglycosylase H16 A0665</td>\n",
       "      <td>395</td>\n",
       "      <td>862</td>\n",
       "      <td>+</td>\n",
       "      <td>37.2</td>\n",
       "      <td>107.8</td>\n",
       "      <td>1.500000e-22</td>\n",
       "      <td>undefined</td>\n",
       "      <td>0.0</td>\n",
       "      <td>undefined</td>\n",
       "      <td>0.675325</td>\n",
       "      <td>0</td>\n",
       "      <td>2392.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>5.402684</td>\n",
       "      <td>2.258647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_16501</th>\n",
       "      <td>BAC0162</td>\n",
       "      <td>Gallium</td>\n",
       "      <td>MRGs</td>\n",
       "      <td>Gallium</td>\n",
       "      <td>15</td>\n",
       "      <td>659</td>\n",
       "      <td>+</td>\n",
       "      <td>31.0</td>\n",
       "      <td>102.8</td>\n",
       "      <td>9.400000e-24</td>\n",
       "      <td>1486262</td>\n",
       "      <td>196.0</td>\n",
       "      <td>Martelella endophytica</td>\n",
       "      <td>0.606742</td>\n",
       "      <td>0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>2.734692</td>\n",
       "      <td>1.568957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold_16501</th>\n",
       "      <td>BAC0604</td>\n",
       "      <td>Molybdenum</td>\n",
       "      <td>MRGs</td>\n",
       "      <td>Molybdenum</td>\n",
       "      <td>778</td>\n",
       "      <td>1440</td>\n",
       "      <td>+</td>\n",
       "      <td>30.0</td>\n",
       "      <td>102.8</td>\n",
       "      <td>9.500000e-24</td>\n",
       "      <td>1486262</td>\n",
       "      <td>196.0</td>\n",
       "      <td>Martelella endophytica</td>\n",
       "      <td>0.648256</td>\n",
       "      <td>0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>2.734692</td>\n",
       "      <td>1.568957</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id  \\\n",
       "read                              \n",
       "scaffold_16508   WP_007306505.1   \n",
       "scaffold_9236      YP_002467958   \n",
       "scaffold_9236   UniRef90_Q07259   \n",
       "scaffold_16501          BAC0162   \n",
       "scaffold_16501          BAC0604   \n",
       "\n",
       "                                                        gene_name  \\\n",
       "read                                                                \n",
       "scaffold_16508  IS1634 family transposase [Crocosphaera watsonii]   \n",
       "scaffold_9236                                              PBP-1B   \n",
       "scaffold_9236        Cluster: Putative transglycosylase H16 A0665   \n",
       "scaffold_16501                                            Gallium   \n",
       "scaffold_16501                                         Molybdenum   \n",
       "\n",
       "                           group  \\\n",
       "read                               \n",
       "scaffold_16508              MGEs   \n",
       "scaffold_9236               ARGs   \n",
       "scaffold_9236   Functional Genes   \n",
       "scaffold_16501              MRGs   \n",
       "scaffold_16501              MRGs   \n",
       "\n",
       "                                                    category  start   end  \\\n",
       "read                                                                        \n",
       "scaffold_16508                                   transposase   1058  1459   \n",
       "scaffold_9236                                    beta-lactam    326  1816   \n",
       "scaffold_9236   Cluster: Putative transglycosylase H16 A0665    395   862   \n",
       "scaffold_16501                                       Gallium     15   659   \n",
       "scaffold_16501                                    Molybdenum    778  1440   \n",
       "\n",
       "               strand  identity  bitscore        evalue NCBI_taxa_id  \\\n",
       "read                                                                   \n",
       "scaffold_16508      +      35.1     102.4  6.500000e-21    undefined   \n",
       "scaffold_9236       +      26.6     137.9  1.300000e-32    undefined   \n",
       "scaffold_9236       +      37.2     107.8  1.500000e-22    undefined   \n",
       "scaffold_16501      +      31.0     102.8  9.400000e-24      1486262   \n",
       "scaffold_16501      +      30.0     102.8  9.500000e-24      1486262   \n",
       "\n",
       "                taxa_centrifuge_score                 species  coverage  \\\n",
       "read                                                                      \n",
       "scaffold_16508                    0.0               undefined  0.724324   \n",
       "scaffold_9236                     0.0               undefined  0.718421   \n",
       "scaffold_9236                     0.0               undefined  0.675325   \n",
       "scaffold_16501                  196.0  Martelella endophytica  0.606742   \n",
       "scaffold_16501                  196.0  Martelella endophytica  0.648256   \n",
       "\n",
       "                is_pathogen  length  read_count  normalized_read_count  \\\n",
       "read                                                                     \n",
       "scaffold_16508            0  1743.0        74.0               4.935786   \n",
       "scaffold_9236             0  2392.0        81.0               5.402684   \n",
       "scaffold_9236             0  2392.0        81.0               5.402684   \n",
       "scaffold_16501            0  1743.0        41.0               2.734692   \n",
       "scaffold_16501            0  1743.0        41.0               2.734692   \n",
       "\n",
       "                normalized_read_count_per_kb  \n",
       "read                                          \n",
       "scaffold_16508                      2.831776  \n",
       "scaffold_9236                       2.258647  \n",
       "scaffold_9236                       2.258647  \n",
       "scaffold_16501                      1.568957  \n",
       "scaffold_16501                      1.568957  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_full_tables_assembly_data['After_GAC_1'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "SAVE DICTIONARY\n",
    "---------------\n",
    "save dictionary in all_full_tables_assembly_data_test.p\n",
    "\n",
    "dictionary can be loaded in new notebook for analysis\n",
    "\n",
    "with open('all_full_tables_remove_overlap_test.p', 'rb') as fp:\n",
    "    all_full_tables_remove_overlap = pickle.load(fp)\n",
    "\"\"\"\n",
    "\n",
    "with open('all_full_tables_assembly_data_test.p', 'wb') as fp:\n",
    "    pickle.dump(all_full_tables_assembly_data, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Running 'remove_overlap_not_ARG_strand_selection'\n",
    "------------------------------------------------\n",
    "rearranging with ascending start\n",
    "and storing updated dataframes in \n",
    "a dictionary called\n",
    "all_full_tables_remove_overlap.\n",
    "\n",
    "Set the max_overlap to desired value.\n",
    "\n",
    "IMPORTANT: this step is necessary for running\n",
    "many of the other functions.\n",
    "\n",
    "\"\"\"\n",
    "all_full_tables_remove_overlap = {}\n",
    "for name, keys in all_full_tables_assembly_data.items():    \n",
    "    df_name = name\n",
    "    # printing sample name indicates progress\n",
    "    print(df_name)\n",
    "    all_full_tables_remove_overlap[df_name] = (\n",
    "        remove_overlap_not_ARG_strand_selection(all_full_tables_assembly_data[df_name], \n",
    "                                            max_overlap=250, \n",
    "                                            strand_selection='no',\n",
    "                                            verbose = False))\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "SAVE DICTIONARY\n",
    "---------------\n",
    "\n",
    "Helpful if you have a lot of samples\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "with open('all_full_tables_remove_overlap_test.p', 'wb') as fp:\n",
    "    pickle.dump(all_full_tables_remove_overlap, fp, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "runs 'count_scaffolds'\n",
    "---------------------\n",
    "\n",
    "Returns dataframe with number of scaffolds in a sample.\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "scaffold_count = pd.DataFrame(dtype=float)\n",
    "for name, df in assembly.items():    \n",
    "    count = count_scaffolds(df, min_length = 0)\n",
    "    count_series = pd.Series(count)\n",
    "    count_rename = count_series.rename(name)\n",
    "    \n",
    "    scaffold_count = scaffold_count.append(count_rename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># scaffolds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>After_GAC_1</th>\n",
       "      <td>304503.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Before_GAC_1</th>\n",
       "      <td>257626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2</th>\n",
       "      <td>216141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W3</th>\n",
       "      <td>235136.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              # scaffolds\n",
       "After_GAC_1      304503.0\n",
       "Before_GAC_1     257626.0\n",
       "HCL0_W2          216141.0\n",
       "HCL0_W3          235136.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scaffold_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After_GAC_1\n",
      "Before_GAC_1\n",
      "HCL0_W2\n",
      "HCL0_W3\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "runs 'count_scaffolds_with_MGEs_and_ARGs_normalized'\n",
    "----------------------------------------------------\n",
    "\n",
    "Necessary to run 'remove_overlap_not_ARG_strand_selection' and 'count_scaffolds' before\n",
    "\n",
    "Estimate: 5 samples run in ~ 2 mins to run. More samples will take longer. \n",
    "\n",
    "\n",
    "Output\n",
    "------\n",
    "df : pd.dataframe\n",
    "    dataframe with counts, ratios and percent values\n",
    "    \n",
    "    ARG_counts : Number of ARGs in a sample \n",
    "    scaffold_ARGs : Number of scaffolds with at least one ARG \n",
    "    scaffold_ARGs_normalized : Scaffold with at least one ARG \n",
    "                               rel. abundance (rpkm)\n",
    "    scaffold_MGEs : Number of scaffolds with at least one MGE\n",
    "    scaffold_MGEs_with_ARGs : Number of scaffolds with at least one AGR \n",
    "                              and at least one MGE rel. abundance (rpkm)\n",
    "    scaffold_MGEs_with_ARGs_normalized : Scaffold with at least one each \n",
    "                                         of ARG and MGE\n",
    "    %_scaffold_MGEs_with_ARGs/scaffold_ARGs : (scaffold_MGEs_with_ARGs x 100)/scaffold_ARGs\n",
    "    scaffold_pathogens : Number of scaffolds identified as pathogen\n",
    "    scaffold_pathogens_with_ARGs : Number of scaffolds identified as pathogen with \n",
    "                                at least one ARG\n",
    "    scaffold_pathogens_with_MGEs : Number of scaffolds identified as pathogen with \n",
    "                                at least one MGE\n",
    "    scaffold_pathogens_with_ARGs_MGEs : Number of scaffolds identified \n",
    "                                        as pathogen with at least one AGR \n",
    "                                        and at least one MGE\n",
    "    scaffold_has_pathogens_ARGs_MGEs_normalized: Rel abundance of scaffolds identified \n",
    "                                        as pathogen with at least one AGR \n",
    "                                        and at least one MGE (rpkm)\n",
    "    # scaffolds :  Number of scaffolds in a sample\n",
    "    scaffold_MGEs/#scaffolds : scaffold_MGEs/# scaffolds\n",
    "    %scaffold_ARGs : (scaffold_ARGs  x 100) / # scaffolds\n",
    "    %scaffold_MGEs_with_ARGs : (scaffold_MGEs_with_ARGs  x 100) / # scaffolds \n",
    "    %pathogens : (scaffold_pathogens x 100) / # scaffolds \n",
    "    %pathogens_with_ARGs : (scaffold_pathogens_with_ARGs x 100) / # scaffolds\n",
    "    %pathogens_with_MGEs : (scaffold_pathogens_with_MGEs x 100) / # scaffolds\n",
    "    %pathogens_with_ARGs_MGEs : (scaffold_pathogens_with_ARGs_MGEs x 100) / # scaffolds \n",
    "    %scaffold_MGEs : (scaffold_MGEs x 100) / # scaffolds\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "columns = [\n",
    "           'ARG_counts', 'scaffold_ARGs',\n",
    "           'scaffold_ARGs_normalized',\n",
    "           'scaffold_MGEs',\n",
    "           'scaffold_MGEs_with_ARGs',\n",
    "           'scaffold_MGEs_with_ARGs_normalized',\n",
    "           '%_scaffold_MGEs_with_ARGs/scaffold_ARGs', 'scaffold_pathogens',\n",
    "           'scaffold_pathogens_with_ARGs', 'scaffold_pathogens_with_MGEs',\n",
    "           'scaffold_pathogens_with_ARGs_MGEs',\n",
    "           'scaffold_has_pathogens_ARGs_MGEs_normalized']\n",
    "scaffolds_with_MGEs_and_ARGs = pd.DataFrame(columns=columns, dtype=float)\n",
    "for name, df in all_full_tables_remove_overlap.items(): \n",
    "    print(name)\n",
    "    count = count_scaffolds_with_MGEs_and_ARGs_normalized(df, min_length = 500)\n",
    "    count_series = pd.Series(count)\n",
    "    count_rename = count_series.rename(name)\n",
    "    \n",
    "    scaffolds_with_MGEs_and_ARGs = scaffolds_with_MGEs_and_ARGs.append(count_rename)\n",
    "        \n",
    "\n",
    "scaffolds_with_MGEs_and_ARGs_store = scaffolds_with_MGEs_and_ARGs\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized = (\n",
    "    pd.concat([scaffolds_with_MGEs_and_ARGs_store, scaffold_count], \n",
    "              axis=1, join='inner'))\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs/#scaffolds']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs']/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%scaffold_ARGs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_ARGs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%scaffold_MGEs_with_ARGs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs_with_ARGs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens_with_ARGs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens_with_ARGs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens_with_MGEs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens_with_MGEs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%pathogens_with_ARGs_MGEs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_pathogens_with_ARGs_MGEs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized['%scaffold_MGEs']=\\\n",
    "MGEs_and_ARGs_pathogens_normalized['scaffold_MGEs']*100/\\\n",
    "MGEs_and_ARGs_pathogens_normalized['# scaffolds']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARG_counts</th>\n",
       "      <th>scaffold_ARGs</th>\n",
       "      <th>scaffold_ARGs_normalized</th>\n",
       "      <th>scaffold_MGEs</th>\n",
       "      <th>scaffold_MGEs_with_ARGs</th>\n",
       "      <th>scaffold_MGEs_with_ARGs_normalized</th>\n",
       "      <th>%_scaffold_MGEs_with_ARGs/scaffold_ARGs</th>\n",
       "      <th>scaffold_pathogens</th>\n",
       "      <th>scaffold_pathogens_with_ARGs</th>\n",
       "      <th>scaffold_pathogens_with_MGEs</th>\n",
       "      <th>...</th>\n",
       "      <th>scaffold_has_pathogens_ARGs_MGEs_normalized</th>\n",
       "      <th># scaffolds</th>\n",
       "      <th>scaffold_MGEs/#scaffolds</th>\n",
       "      <th>%scaffold_ARGs</th>\n",
       "      <th>%scaffold_MGEs_with_ARGs</th>\n",
       "      <th>%pathogens</th>\n",
       "      <th>%pathogens_with_ARGs</th>\n",
       "      <th>%pathogens_with_MGEs</th>\n",
       "      <th>%pathogens_with_ARGs_MGEs</th>\n",
       "      <th>%scaffold_MGEs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>After_GAC_1</th>\n",
       "      <td>5068.0</td>\n",
       "      <td>3974.0</td>\n",
       "      <td>13837.251645</td>\n",
       "      <td>3075.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>3212.788846</td>\n",
       "      <td>9.788626</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>304503.0</td>\n",
       "      <td>0.010098</td>\n",
       "      <td>1.305077</td>\n",
       "      <td>0.127749</td>\n",
       "      <td>0.000985</td>\n",
       "      <td>0.000328</td>\n",
       "      <td>0.000657</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.009842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Before_GAC_1</th>\n",
       "      <td>4130.0</td>\n",
       "      <td>3424.0</td>\n",
       "      <td>17877.630338</td>\n",
       "      <td>2425.0</td>\n",
       "      <td>205.0</td>\n",
       "      <td>3148.961065</td>\n",
       "      <td>5.987150</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>72.295621</td>\n",
       "      <td>257626.0</td>\n",
       "      <td>0.009413</td>\n",
       "      <td>1.329058</td>\n",
       "      <td>0.079573</td>\n",
       "      <td>0.006599</td>\n",
       "      <td>0.003493</td>\n",
       "      <td>0.003882</td>\n",
       "      <td>0.000776</td>\n",
       "      <td>0.941287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2</th>\n",
       "      <td>6325.0</td>\n",
       "      <td>4894.0</td>\n",
       "      <td>19977.884597</td>\n",
       "      <td>3447.0</td>\n",
       "      <td>463.0</td>\n",
       "      <td>5422.032241</td>\n",
       "      <td>9.460564</td>\n",
       "      <td>31.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>...</td>\n",
       "      <td>22.399722</td>\n",
       "      <td>216141.0</td>\n",
       "      <td>0.015948</td>\n",
       "      <td>2.264263</td>\n",
       "      <td>0.214212</td>\n",
       "      <td>0.014342</td>\n",
       "      <td>0.003239</td>\n",
       "      <td>0.010179</td>\n",
       "      <td>0.000925</td>\n",
       "      <td>1.594792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W3</th>\n",
       "      <td>7336.0</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>20610.972909</td>\n",
       "      <td>3938.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>5527.301233</td>\n",
       "      <td>9.660897</td>\n",
       "      <td>30.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>26.447524</td>\n",
       "      <td>235136.0</td>\n",
       "      <td>0.016748</td>\n",
       "      <td>2.407968</td>\n",
       "      <td>0.232631</td>\n",
       "      <td>0.012759</td>\n",
       "      <td>0.002126</td>\n",
       "      <td>0.010632</td>\n",
       "      <td>0.000851</td>\n",
       "      <td>1.674775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ARG_counts  scaffold_ARGs  scaffold_ARGs_normalized  \\\n",
       "After_GAC_1       5068.0         3974.0              13837.251645   \n",
       "Before_GAC_1      4130.0         3424.0              17877.630338   \n",
       "HCL0_W2           6325.0         4894.0              19977.884597   \n",
       "HCL0_W3           7336.0         5662.0              20610.972909   \n",
       "\n",
       "              scaffold_MGEs  scaffold_MGEs_with_ARGs  \\\n",
       "After_GAC_1          3075.0                    389.0   \n",
       "Before_GAC_1         2425.0                    205.0   \n",
       "HCL0_W2              3447.0                    463.0   \n",
       "HCL0_W3              3938.0                    547.0   \n",
       "\n",
       "              scaffold_MGEs_with_ARGs_normalized  \\\n",
       "After_GAC_1                          3212.788846   \n",
       "Before_GAC_1                         3148.961065   \n",
       "HCL0_W2                              5422.032241   \n",
       "HCL0_W3                              5527.301233   \n",
       "\n",
       "              %_scaffold_MGEs_with_ARGs/scaffold_ARGs  scaffold_pathogens  \\\n",
       "After_GAC_1                                  9.788626                 3.0   \n",
       "Before_GAC_1                                 5.987150                17.0   \n",
       "HCL0_W2                                      9.460564                31.0   \n",
       "HCL0_W3                                      9.660897                30.0   \n",
       "\n",
       "              scaffold_pathogens_with_ARGs  scaffold_pathogens_with_MGEs  ...  \\\n",
       "After_GAC_1                            1.0                           2.0  ...   \n",
       "Before_GAC_1                           9.0                          10.0  ...   \n",
       "HCL0_W2                                7.0                          22.0  ...   \n",
       "HCL0_W3                                5.0                          25.0  ...   \n",
       "\n",
       "              scaffold_has_pathogens_ARGs_MGEs_normalized  # scaffolds  \\\n",
       "After_GAC_1                                      0.000000     304503.0   \n",
       "Before_GAC_1                                    72.295621     257626.0   \n",
       "HCL0_W2                                         22.399722     216141.0   \n",
       "HCL0_W3                                         26.447524     235136.0   \n",
       "\n",
       "              scaffold_MGEs/#scaffolds  %scaffold_ARGs  \\\n",
       "After_GAC_1                   0.010098        1.305077   \n",
       "Before_GAC_1                  0.009413        1.329058   \n",
       "HCL0_W2                       0.015948        2.264263   \n",
       "HCL0_W3                       0.016748        2.407968   \n",
       "\n",
       "              %scaffold_MGEs_with_ARGs  %pathogens  %pathogens_with_ARGs  \\\n",
       "After_GAC_1                   0.127749    0.000985              0.000328   \n",
       "Before_GAC_1                  0.079573    0.006599              0.003493   \n",
       "HCL0_W2                       0.214212    0.014342              0.003239   \n",
       "HCL0_W3                       0.232631    0.012759              0.002126   \n",
       "\n",
       "              %pathogens_with_MGEs  %pathogens_with_ARGs_MGEs  %scaffold_MGEs  \n",
       "After_GAC_1               0.000657                   0.000000        1.009842  \n",
       "Before_GAC_1              0.003882                   0.000776        0.941287  \n",
       "HCL0_W2                   0.010179                   0.000925        1.594792  \n",
       "HCL0_W3                   0.010632                   0.000851        1.674775  \n",
       "\n",
       "[4 rows x 21 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Prints result.\n",
    "Save this file and plot using desired software\n",
    "\"\"\"\n",
    "\n",
    "MGEs_and_ARGs_pathogens_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "runs 'ARG_subtype_count'\n",
    "-----------------------\n",
    "\n",
    "\n",
    "Run 'scaffold_count' before with desired scaffold length threshhold.\n",
    "\n",
    "Returns a dataframe with count of ARG subtype.\n",
    "\"\"\"\n",
    "\n",
    "columns = ['arg_subtype_count', 'scaffold_arg_subtype', \n",
    "           'scaffold_has_arg_subtype_normalized']\n",
    "ARG_subtypes = pd.DataFrame(columns=columns, dtype=float)\n",
    "for name, df in all_full_tables_assembly_data.items():    \n",
    "    count = ARG_subtype_count(df, min_bitscore=50, \n",
    "                      min_id_arg=50, min_coverage_arg = 0.5, arg_name = 'sul1')\n",
    "    count_series = pd.Series(count)\n",
    "    count_rename = count_series.rename(name)\n",
    "    \n",
    "    ARG_subtypes = ARG_subtypes.append(count_rename)\n",
    "        \n",
    "\n",
    "ARG_subtypes_store = ARG_subtypes \n",
    "\n",
    "\n",
    "ARG_subtypes_scaffold_count = \\\n",
    "pd.concat([ARG_subtypes_store, scaffold_count], axis=1, join='inner')\n",
    "\n",
    "ARG_subtypes_scaffold_count['%scaffold_arg_subtype']=\\\n",
    "ARG_subtypes_scaffold_count['scaffold_arg_subtype']*100/\\\n",
    "ARG_subtypes_scaffold_count['# scaffolds']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>arg_subtype_count</th>\n",
       "      <th>scaffold_arg_subtype</th>\n",
       "      <th>scaffold_has_arg_subtype_normalized</th>\n",
       "      <th># scaffolds</th>\n",
       "      <th>%scaffold_arg_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>After_GAC_1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>304503.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Before_GAC_1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.291695</td>\n",
       "      <td>257626.0</td>\n",
       "      <td>0.000388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17.712378</td>\n",
       "      <td>216141.0</td>\n",
       "      <td>0.000463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCL0_W3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.809571</td>\n",
       "      <td>235136.0</td>\n",
       "      <td>0.000425</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              arg_subtype_count  scaffold_arg_subtype  \\\n",
       "After_GAC_1                 0.0                   0.0   \n",
       "Before_GAC_1                1.0                   1.0   \n",
       "HCL0_W2                     1.0                   1.0   \n",
       "HCL0_W3                     1.0                   1.0   \n",
       "\n",
       "              scaffold_has_arg_subtype_normalized  # scaffolds  \\\n",
       "After_GAC_1                              0.000000     304503.0   \n",
       "Before_GAC_1                            13.291695     257626.0   \n",
       "HCL0_W2                                 17.712378     216141.0   \n",
       "HCL0_W3                                 14.809571     235136.0   \n",
       "\n",
       "              %scaffold_arg_subtype  \n",
       "After_GAC_1                0.000000  \n",
       "Before_GAC_1               0.000388  \n",
       "HCL0_W2                    0.000463  \n",
       "HCL0_W3                    0.000425  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ARG_subtypes_scaffold_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
